From 85e2802fa7f1818f6b064bd954079eb283df812a Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Tue, 16 Dec 2014 19:22:41 +0900
Subject: [PATCH 1/7] .gitignore: Add version script.

---
 .gitignore | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/.gitignore b/.gitignore
index d9cf6d5..ae1cc88 100644
--- a/.gitignore
+++ b/.gitignore
@@ -14,6 +14,8 @@
 *.rej
 *.so*
 *.swp
+*.v
+*.ver
 .depend
 .DS_Store
 config.*
-- 
2.2.0

