From 315bbfa8acb5dcfda9e34711255eef9a5f9eacc8 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 14 Sep 2014 05:35:42 +0900
Subject: [PATCH 2/3] configure: Add api version to SHAREDLIBNAME.

---
 configure | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/configure b/configure
index a28028a..35b570b 100755
--- a/configure
+++ b/configure
@@ -228,15 +228,6 @@ case "$TARGET_OS" in
 esac
 
 
-STATICLIBNAME="${STATIC_NAME}${STATIC_EXT}"
-SHAREDLIBNAME="${SHARED_NAME}${SHARED_EXT}"
-test -n "$STATICLIB" && STATICLIB="$STATICLIBNAME"
-test -n "$SHAREDLIB" && SHAREDLIB="$SHAREDLIBNAME"
-test -z "$STATICLIB" -a -z "$SHAREDLIB" && \
-    error_exit "--disable-static requires --enable-shared were specified"
-test -z "$SHAREDLIB" && SO_LDFLAGS=""
-
-
 CFLAGS="$CFLAGS $XCFLAGS"
 LDFLAGS="$LDFLAGS $XLDFLAGS"
 LIBS="$LIBS $XLIBS"
@@ -312,6 +303,14 @@ cat >> config.h << EOF
 EOF
 
 
+STATICLIBNAME="${STATIC_NAME}${STATIC_EXT}"
+SHAREDLIBNAME="${SHARED_NAME}-${MAJVER}${SHARED_EXT}"
+test -n "$STATICLIB" && STATICLIB="$STATICLIBNAME"
+test -n "$SHAREDLIB" && SHAREDLIB="$SHAREDLIBNAME"
+test -z "$STATICLIB" -a -z "$SHAREDLIB" && \
+    error_exit "--disable-static requires --enable-shared were specified"
+test -z "$SHAREDLIB" && SO_LDFLAGS=""
+
 cat >> liblsmash.pc << EOF
 prefix=$prefix
 exec_prefix=$exec_prefix
-- 
2.1.0

