From 6b0265d803b50161448b4cec09bfa0e172dcc99b Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 28 Sep 2014 15:26:19 +0900
Subject: [PATCH 3/3] build: Use lib.exe when it is available on mingw.

---
 .gitignore |  1 +
 Makefile   |  9 ++++++++-
 configure  | 26 ++++++++++++++++++++++++++
 3 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/.gitignore b/.gitignore
index a20f4c2..9ae81bd 100644
--- a/.gitignore
+++ b/.gitignore
@@ -6,6 +6,7 @@
 *.dll*
 *.dylib
 *.exe
+*.exp
 *.lib
 *.o
 *.orig
diff --git a/Makefile b/Makefile
index 4f7a7a5..8b1f8f7 100644
--- a/Makefile
+++ b/Makefile
@@ -28,6 +28,9 @@ $(STATICLIBNAME): $(OBJS)
 $(SHAREDLIBNAME): $(OBJS)
 	$(LD) $(SO_LDFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS)
 	-@ $(if $(STRIP), $(STRIP) -x $@)
+ifneq ($(SLIB_CMD),)
+	$(SLIB_CMD)
+endif
 ifeq ($(SHAREDLIBNAME), liblsmash.so.$(MAJVER))
 	ln -s $(SHAREDLIBNAME) liblsmash.so
 endif
@@ -59,6 +62,10 @@ ifneq ($(IMPLIB),)
 	install -m 644 $(IMPLIB) $(DESTDIR)$(libdir)
 	install -d $(DESTDIR)$(bindir)
 	install -m 755 $(SHAREDLIB) $(DESTDIR)$(bindir)
+ifneq ($(SLIB_CMD),)
+	install -m 644 $(DEFNAME) $(DESTDIR)$(libdir)
+	install -m 644 lsmash.lib $(DESTDIR)$(bindir)
+endif
 ifeq ($(SHAREDLIB), liblsmash-$(MAJVER).dll)
 	ln -s $(DESTDIR)$(bindir)/$(SHAREDLIB) $(DESTDIR)$(bindir)/liblsmash.dll
 endif
@@ -77,7 +84,7 @@ uninstall:
 	$(RM) $(addprefix $(DESTDIR)$(bindir)/, $(TOOLS_ALL) $(TOOLS_ALL:%=%.exe) liblsmash.dll cyglsmash.dll)
 
 clean:
-	$(RM) */*.o *.a *.so* *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend
+	$(RM) */*.o *.a *.so* *.def *.exp *.lib *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend
 
 distclean: clean
 	$(RM) config.* *.pc
diff --git a/configure b/configure
index 36e0d41..b196d72 100755
--- a/configure
+++ b/configure
@@ -56,6 +56,14 @@ cc_check()
     return $ret
 }
 
+is_64bit()
+{
+    echo 'int main(void){int a[2*(sizeof(void *)>4)-1]; return 0;}' > conftest.c
+    $CC conftest.c -o conftest 2> /dev/null
+    ret=$?
+    rm -f conftest*
+    return $ret
+}
 #-----------------------------------------------------------------------------
 
 rm -f config.* .depend conftest* liblsmash.pc
@@ -244,6 +252,21 @@ test -z "$STATICLIB" -a -z "$SHAREDLIB" && \
     error_exit "--disable-static requires --enable-shared were specified"
 test -z "$SHAREDLIB" && SO_LDFLAGS=""
 
+case "$TARGET_OS" in
+    *mingw*)
+        if lib.exe --list > /dev/null 2>&1 \
+        && type gendef > /dev/null 2>&1 ; then
+            if is_64bit ; then
+                LIBARCH=x64
+            else
+                LIBARCH=i386
+            fi
+            DEFNAME="lsmash-$MAJVER.def"
+            SLIB_CMD='gendef - $(SHAREDLIB) > $(DEFNAME) 2>/dev/null; lib.exe -machine:$(LIBARCH) -def:$(DEFNAME) -out:lsmash.lib'
+        fi
+        ;;
+esac
+
 
 CFLAGS="$CFLAGS $XCFLAGS"
 LDFLAGS="$LDFLAGS $XLDFLAGS"
@@ -357,6 +380,9 @@ CFLAGS = $CFLAGS
 LDFLAGS = $LDFLAGS
 SO_LDFLAGS = $SO_LDFLAGS
 LIBS = $LIBS
+LIBARCH = $LIBARCH
+DEFNAME = $DEFNAME
+SLIB_CMD = $SLIB_CMD
 EOF
 
 cat config.mak
-- 
2.1.1

