From 5ae633486a72fb6d4e682c90dc1021ae01351a95 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Mon, 27 Oct 2014 16:01:04 +0900
Subject: [PATCH 3/3] build: Use lib.exe or dlltool when available on mingw.

---
 .gitignore |  1 +
 Makefile   |  8 ++++++++
 configure  | 27 ++++++++++++++++++++++++++-
 3 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/.gitignore b/.gitignore
index a20f4c2..9ae81bd 100644
--- a/.gitignore
+++ b/.gitignore
@@ -6,6 +6,7 @@
 *.dll*
 *.dylib
 *.exe
+*.exp
 *.lib
 *.o
 *.orig
diff --git a/Makefile b/Makefile
index bfd2ccc..b7ba5bf 100644
--- a/Makefile
+++ b/Makefile
@@ -28,6 +28,9 @@ $(STATICLIBNAME): $(OBJS)
 $(SHAREDLIBNAME): $(OBJS)
 	$(LD) $(SO_LDFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS)
 	-@ $(if $(STRIP), $(STRIP) -x $@)
+ifneq ($(SLIB_CMD),)
+	$(SLIB_CMD)
+endif
 ifeq ($(SHAREDLIBNAME), liblsmash.so.$(MAJVER))
 	ln -s $(SHAREDLIBNAME) liblsmash.so
 endif
@@ -59,6 +62,10 @@ ifneq ($(IMPLIB),)
 	install -m 644 $(IMPLIB) $(DESTDIR)$(libdir)
 	install -d $(DESTDIR)$(bindir)
 	install -m 755 $(SHAREDLIB) $(DESTDIR)$(bindir)
+ifneq ($(SLIB_CMD),)
+	install -m 644 $(DEFNAME) $(DESTDIR)$(libdir)
+	install -m 644 lsmash.lib $(DESTDIR)$(bindir)
+endif
 ifeq ($(SHAREDLIB), liblsmash-$(MAJVER).dll)
 	ln -fs $(DESTDIR)$(bindir)/$(SHAREDLIB) $(DESTDIR)$(bindir)/liblsmash.dll
 endif
@@ -78,6 +85,7 @@ uninstall:
 
 clean:
 	$(RM) */*.o *.a *.so* *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend
+	$(RM) */*.o *.a *.so* *.def *.exp *.lib *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend
 
 distclean: clean
 	$(RM) config.* *.pc
diff --git a/configure b/configure
index 74036b9..bbdea4e 100755
--- a/configure
+++ b/configure
@@ -56,6 +56,15 @@ cc_check()
     return $ret
 }
 
+is_64bit()
+{
+    echo 'int main(void){int a[2*(sizeof(void *)>4)-1]; return 0;}' > conftest.c
+    $CC conftest.c -o conftest 2> /dev/null
+    ret=$?
+    rm -f conftest*
+    return $ret
+}
+
 #-----------------------------------------------------------------------------
 
 rm -f config.* .depend conftest* liblsmash.pc
@@ -206,9 +215,22 @@ case "$TARGET_OS" in
         EXT=".exe"
         SHARED_NAME="liblsmash-$MAJVER"
         SHARED_EXT=".dll"
+        DEFNAME="${SHARED_NAME}.def"
         IMPLIB="liblsmash.dll.a"
-        SO_LDFLAGS="-shared -Wl,--out-implib,$IMPLIB"
+        SO_LDFLAGS="-shared -Wl,--output-def,$DEFNAME -Wl,--out-implib,$IMPLIB"
         CFLAGS="$CFLAGS -D__USE_MINGW_ANSI_STDIO=1"
+        LIBARCH=i386
+        if lib.exe --list > /dev/null 2>&1 ; then
+            if is_64bit ; then
+                LIBARCH=x64
+            fi
+            SLIB_CMD='sed -i "s/ @[^ ]*//" $(DEFNAME); lib.exe -machine:$(LIBARCH) -def:$(DEFNAME) -out:lsmash.lib'
+        elif ${CROSS}dlltool --version > /dev/null 2>&1 ; then
+            if is_64bit ; then
+                LIBARCH="i386:x86-64"
+            fi
+            SLIB_CMD="sed -i \"s/ @[^ ]*//\" \$(DEFNAME); ${CROSS}dlltool -m \$(LIBARCH) -d \$(DEFNAME) -l lsmash.lib -D \$(SHAREDLIBNAME)"
+        fi
         ;;
     *cygwin*)
         EXT=".exe"
@@ -371,6 +393,9 @@ CFLAGS = $CFLAGS
 LDFLAGS = $LDFLAGS
 SO_LDFLAGS = $SO_LDFLAGS
 LIBS = $LIBS
+LIBARCH = $LIBARCH
+DEFNAME = $DEFNAME
+SLIB_CMD = $SLIB_CMD
 EOF
 
 cat config.mak
-- 
2.1.2

