From bb0282ddb596352b83365873713f06463f1cdca6 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Tue, 2 Sep 2014 01:22:52 +0900
Subject: [PATCH 3/4] configure: Add api version to SHAREDLIBNAME.

---
 configure | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/configure b/configure
index bc4840c..d921f80 100755
--- a/configure
+++ b/configure
@@ -228,19 +228,6 @@ case "$TARGET_OS" in
 esac
 
 
-STATICLIBNAME="${STATIC_NAME}${STATIC_EXT}"
-SHAREDLIBNAME="${SHARED_NAME}${SHARED_EXT}"
-test -n "$STATICLIB" && STATICLIB="$STATICLIBNAME"
-test -n "$SHAREDLIB" && SHAREDLIB="$SHAREDLIBNAME"
-test -z "$STATICLIB" -a -z "$SHAREDLIB" && \
-    error_exit "--disable-static requires --enable-shared were specified"
-test -z "$SHAREDLIB" && SO_LDFLAGS=""
-if test -z "$STATICLIB" -a -n "$SHAREDLIB"; then
-    CLI_LIBLSMASH="$SHAREDLIBNAME"
-else
-    CLI_LIBLSMASH="$STATICLIBNAME"
-fi
-
 CFLAGS="$CFLAGS $XCFLAGS"
 LDFLAGS="$LDFLAGS $XLDFLAGS"
 LIBS="$LIBS $XLIBS"
@@ -316,6 +303,19 @@ cat >> config.h << EOF
 EOF
 
 
+STATICLIBNAME="${STATIC_NAME}${STATIC_EXT}"
+SHAREDLIBNAME="${SHARED_NAME}-${MAJVER}${SHARED_EXT}"
+test -n "$STATICLIB" && STATICLIB="$STATICLIBNAME"
+test -n "$SHAREDLIB" && SHAREDLIB="$SHAREDLIBNAME"
+test -z "$STATICLIB" -a -z "$SHAREDLIB" && \
+    error_exit "--disable-static requires --enable-shared were specified"
+test -z "$SHAREDLIB" && SO_LDFLAGS=""
+if test -z "$STATICLIB" -a -n "$SHAREDLIB"; then
+    CLI_LIBLSMASH="$SHAREDLIBNAME"
+else
+    CLI_LIBLSMASH="$STATICLIBNAME"
+fi
+
 cat >> liblsmash.pc << EOF
 prefix=$prefix
 exec_prefix=$exec_prefix
-- 
2.1.0

