--- a/compiler.h
+++ b/compiler.h
@@ -195,7 +195,11 @@
  * How to tell the compiler that a function takes a printf-like string
  */
 #ifdef __GNUC__
-# define printf_func(fmt, list) __attribute__((format(printf, fmt, list)))
+# ifdef __MINGW32__
+#  define printf_func(fmt, list) __attribute__((format(__MINGW_PRINTF_FORMAT, fmt, list)))
+# else
+#  define printf_func(fmt, list) __attribute__((format(printf, fmt, list)))
+# endif
 #else
 # define printf_func(fmt, list)
 #endif
