From 736152c70e5de92fba61749a11a4e07711a4dd84 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 15 Sep 2013 10:35:05 +0900
Subject: [PATCH] configure: Change mingw32* into mingw*.

---
 configure | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/configure b/configure
index 8d393b0..e35c998 100755
--- a/configure
+++ b/configure
@@ -2632,7 +2632,7 @@ fi
 
 exesuf() {
     case $1 in
-        mingw32*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
+        mingw*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
     esac
 }
 
@@ -3510,11 +3510,11 @@ case $target_os in
         enabled_any pic shared ||
             { check_cflags -mdynamic-no-pic && add_asflags -mdynamic-no-pic; }
         ;;
-    mingw32*)
+    mingw*)
         if test $target_os = "mingw32ce"; then
             disable network
         else
-            target_os=mingw32
+            target_os=mingw
         fi
         LIBTARGET=i386
         if enabled x86_64; then
-- 
1.8.3.msysgit.0

