diff --git a/update_version b/update_version
index a999991..1ae7ab7 100755
--- a/update_version
+++ b/update_version
@@ -34,7 +34,7 @@ fi
 # -dirty from files that have been touched but are not actually altered in the
 # working dir.
 GIT_VERSION=$(cd "$SRCDIR" && git status > /dev/null 2>&1 \
-                           && git describe --tags --match 'v*' --dirty 2> /dev/null)
+                           && git describe --tags --match 'v1*' 2> /dev/null)
 GIT_VERSION=${GIT_VERSION#v}
 
 if [ -n "$GIT_VERSION" ]; then
