From 319848c6eb1ba6f0fe80a10cde9a640804765429 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 28 Sep 2014 04:10:10 +0900
Subject: [PATCH 4/8] configure: Add api version to mingw shared library name.

---
 configure | 1 +
 1 file changed, 1 insertion(+)

diff --git a/configure b/configure
index c748300..220c8dd 100755
--- a/configure
+++ b/configure
@@ -198,6 +198,7 @@ fi
 case "$TARGET_OS" in
     *mingw*)
         EXT=".exe"
+        SHARED_NAME="liblsmash-$MAJVER"
         SHARED_EXT=".dll"
         IMPLIB="liblsmash.dll.a"
         SO_LDFLAGS="-shared -Wl,--out-implib,$IMPLIB -Wl,--version-script,liblsmash.ver"
-- 
2.4.3

