From 803cc7a133ea73b3a5cf5b3eb67addd363fd6b3b Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 9 Nov 2014 15:47:24 +0900
Subject: [PATCH 5/7] configure: If shared is enabled, put $LIBS on
 Libs.private.

---
 configure | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 9d96bc3..9d42e68 100755
--- a/configure
+++ b/configure
@@ -421,8 +421,8 @@ Description: Loyal to Spec of MPEG4, and Ad-hock Simple Hackwork
 Version: $VER_STRING
 Requires:
 URL: https://github.com/l-smash/l-smash
-Libs: -L${libdir} -llsmash $LIBS
-Libs.private:
+Libs: -L${libdir} -llsmash $(test -z "$SHAREDLIB" && echo $LIBS)
+Libs.private: $(test -n "$SHAREDLIB" && echo $LIBS)
 Cflags: -I${includedir}
 EOF
 
-- 
2.2.0

