From 7a7c7f568ab25d356a6b931628147825f582b328 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 9 Nov 2014 15:47:24 +0900
Subject: [PATCH 6/8] configure: If shared is enabled, put $LIBS on
 Libs.private.

---
 configure | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index af6aa8f..3f24182 100755
--- a/configure
+++ b/configure
@@ -420,8 +420,8 @@ Description: Loyal to Spec of MPEG4, and Ad-hock Simple Hackwork
 Version: $VER_STRING
 Requires:
 URL: https://github.com/l-smash/l-smash
-Libs: -L${libdir} -llsmash $LIBS
-Libs.private:
+Libs: -L${libdir} -llsmash $(test -z "$SHAREDLIB" && echo $LIBS)
+Libs.private: $(test -n "$SHAREDLIB" && echo $LIBS)
 Cflags: -I${includedir}
 EOF
 
-- 
2.4.3

