From 44f704873d8a75c0fc37e65de4bcb4509afc4539 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Tue, 16 Dec 2014 20:28:30 +0900
Subject: [PATCH 7/8] Makefile: Move *.ver from clean to distclean.

---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 6846f2f..1532c40 100644
--- a/Makefile
+++ b/Makefile
@@ -78,10 +78,10 @@ uninstall:
 	$(RM) $(addprefix $(DESTDIR)$(bindir)/, $(TOOLS_ALL) $(TOOLS_ALL:%=%.exe) liblsmash.dll cyglsmash.dll)
 
 clean:
-	$(RM) */*.o *.a *.so* *.def *.exp *.lib *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend *.ver
+	$(RM) */*.o *.a *.so* *.def *.exp *.lib *.dll *.dylib $(addprefix cli/, *.exe $(TOOLS_ALL)) .depend
 
 distclean: clean
-	$(RM) config.* *.pc
+	$(RM) config.* *.pc *.ver
 
 dep: .depend
 
-- 
2.4.3

