diff --git a/build/make/Makefile b/build/make/Makefile
index ed90397..a2aff53 100644
--- a/build/make/Makefile
+++ b/build/make/Makefile
@@ -299,7 +299,7 @@ $(1):
 	$(qexec)$$(LD) -shared $$(LDFLAGS) \
             -Wl,--no-undefined -Wl,-soname,$$(SONAME) \
             -Wl,--version-script,$$(EXPORTS_FILE) -o $$@ \
-            $$(filter %.o,$$^) $$(extralibs)
+            -Wl,--out-implib=$$(subst $(2),.dll.a,$(1)) $$(filter %.o,$$^) $$(extralibs)
 endef
 
 define dl_template
@@ -383,7 +383,7 @@ LIBS=$(call enabled,LIBS)
 .libs: $(LIBS)
 	@touch $@
 $(foreach lib,$(filter %_g.a,$(LIBS)),$(eval $(call archive_template,$(lib))))
-$(foreach lib,$(filter %so.$(SO_VERSION_MAJOR).$(SO_VERSION_MINOR).$(SO_VERSION_PATCH),$(LIBS)),$(eval $(call so_template,$(lib))))
+$(foreach lib,$(filter %-$(SO_VERSION_MAJOR).dll,$(LIBS)),$(eval $(call so_template,$(lib),-$(SO_VERSION_MAJOR).dll)))
 $(foreach lib,$(filter %$(SO_VERSION_MAJOR).dylib,$(LIBS)),$(eval $(call dl_template,$(lib))))
 
 INSTALL-LIBS=$(call cond_enabled,CONFIG_INSTALL_LIBS,INSTALL-LIBS)
diff --git a/build/make/configure.sh b/build/make/configure.sh
index 56e9f44..176f2cc 100644
--- a/build/make/configure.sh
+++ b/build/make/configure.sh
@@ -569,11 +569,14 @@ process_common_cmdline() {
       --libdir=*)
         libdir="${optval}"
         ;;
+      --bindir=*)
+        bindir="${optval}"
+        ;;
       --sdk-path=*)
         [ -d "${optval}" ] || die "Not a directory: ${optval}"
         sdk_path="${optval}"
         ;;
-      --libc|--as|--prefix|--libdir|--sdk-path)
+      --libc|--as|--prefix|--libdir|--bindir|--sdk-path)
         die "Option ${opt} requires argument"
         ;;
       --help|-h) show_help
@@ -600,9 +603,14 @@ post_process_common_cmdline() {
   prefix="${prefix%/}"
   libdir="${libdir:-${prefix}/lib}"
   libdir="${libdir%/}"
+  bindir="${bindir:-${prefix}/bin}"
+  bindir="${bindir%/}"
   if [ "${libdir#${prefix}}" = "${libdir}" ]; then
     die "Libdir ${libdir} must be a subdirectory of ${prefix}"
   fi
+  if [ "${bindir#${prefix}}" = "${bindir}" ]; then
+    die "Bindir ${bindir} must be a subdirectory of ${prefix}"
+  fi
 }
 
 
diff --git a/configure b/configure
index 3ed976c..4013216 100755
--- a/configure
+++ b/configure
@@ -511,6 +511,7 @@ else
 DIST_DIR?=\$(DESTDIR)${prefix}
 endif
 LIBSUBDIR=${libdir##${prefix}/}
+BINSUBDIR=${bindir##${prefix}/}
 
 VERSION_STRING=${VERSION_STRING}
 
@@ -544,9 +545,13 @@ process_detect() {
         # magic happens after the command line has been parsed.
         if ! enabled linux; then
             if enabled gnu; then
-                echo "--enable-shared is only supported on ELF; assuming this is OK"
+                echo "--enable-shared is only supported on ELF and PE; assuming this is OK"
+            elif enabled win32; then
+                echo "--enable-shared is only supported on ELF and PE; assuming this is OK"
+            elif enabled win64; then
+                echo "--enable-shared is only supported on ELF and PE; assuming this is OK"
             else
-                die "--enable-shared only supported on ELF for now"
+                die "--enable-shared only supported on ELF and PE for now"
             fi
         fi
     fi
diff --git a/examples.mk b/examples.mk
index fd67a44..23d060a 100644
--- a/examples.mk
+++ b/examples.mk
@@ -275,7 +275,7 @@ BINS-$(NOT_MSVS)           += $(addprefix $(BUILD_PFX),$(ALL_EXAMPLES:.c=$(EXE_S
 
 # Instantiate linker template for all examples.
 CODEC_LIB=$(if $(CONFIG_DEBUG_LIBS),vpx_g,vpx)
-SHARED_LIB_SUF=$(if $(filter darwin%,$(TGT_OS)),.dylib,.so)
+SHARED_LIB_SUF=$(if $(filter darwin%,$(TGT_OS)),.dylib,.dll)
 CODEC_LIB_SUF=$(if $(CONFIG_SHARED),$(SHARED_LIB_SUF),.a)
 $(foreach bin,$(BINS-yes),\
     $(if $(BUILD_OBJS),$(eval $(bin):\
diff --git a/libs.mk b/libs.mk
index f9f2d80..39b9de2 100644
--- a/libs.mk
+++ b/libs.mk
@@ -157,6 +157,7 @@ endif
 INSTALL_MAPS += include/vpx/% $(SRC_PATH_BARE)/vpx/%
 INSTALL_MAPS += include/vpx/% $(SRC_PATH_BARE)/vpx_ports/%
 INSTALL_MAPS += $(LIBSUBDIR)/%     %
+INSTALL_MAPS += $(BINSUBDIR)/%     %
 INSTALL_MAPS += src/%     $(SRC_PATH_BARE)/%
 ifeq ($(CONFIG_MSVS),yes)
 INSTALL_MAPS += $(foreach p,$(VS_PLATFORMS),$(LIBSUBDIR)/$(p)/%  $(p)/Release/%)
@@ -193,6 +194,7 @@ INSTALL-LIBS-$(CONFIG_SHARED) += $(foreach p,$(VS_PLATFORMS),$(LIBSUBDIR)/$(p)/v
 endif
 else
 INSTALL-LIBS-$(CONFIG_STATIC) += $(LIBSUBDIR)/libvpx.a
+INSTALL-LIBS-$(CONFIG_SHARED) += $(LIBSUBDIR)/libvpx.dll.a
 INSTALL-LIBS-$(CONFIG_DEBUG_LIBS) += $(LIBSUBDIR)/libvpx_g.a
 endif
 
@@ -282,18 +284,14 @@ EXPORT_FILE             := libvpx.syms
 LIBVPX_SO_SYMLINKS      := $(addprefix $(LIBSUBDIR)/, \
                              libvpx.dylib  )
 else
-LIBVPX_SO               := libvpx.so.$(SO_VERSION_MAJOR).$(SO_VERSION_MINOR).$(SO_VERSION_PATCH)
+LIBVPX_SO               := libvpx-$(SO_VERSION_MAJOR).dll
 EXPORT_FILE             := libvpx.ver
-LIBVPX_SO_SYMLINKS      := $(addprefix $(LIBSUBDIR)/, \
-                             libvpx.so libvpx.so.$(SO_VERSION_MAJOR) \
-                             libvpx.so.$(SO_VERSION_MAJOR).$(SO_VERSION_MINOR))
 endif
 
-LIBS-$(CONFIG_SHARED) += $(BUILD_PFX)$(LIBVPX_SO)\
-                           $(notdir $(LIBVPX_SO_SYMLINKS))
+LIBS-$(CONFIG_SHARED) += $(BUILD_PFX)$(LIBVPX_SO)
 $(BUILD_PFX)$(LIBVPX_SO): $(LIBVPX_OBJS) $(EXPORT_FILE)
 $(BUILD_PFX)$(LIBVPX_SO): extralibs += -lm
-$(BUILD_PFX)$(LIBVPX_SO): SONAME = libvpx.so.$(SO_VERSION_MAJOR)
+$(BUILD_PFX)$(LIBVPX_SO): SONAME = libvpx-$(SO_VERSION_MAJOR).dll
 $(BUILD_PFX)$(LIBVPX_SO): EXPORTS_FILE = $(EXPORT_FILE)
 
 libvpx.ver: $(call enabled,CODEC_EXPORTS)
@@ -310,7 +307,7 @@ libvpx.syms: $(call enabled,CODEC_EXPORTS)
 CLEAN-OBJS += libvpx.syms
 
 define libvpx_symlink_template
-$(1): $(2)
+$(1): $(3)
 	@echo "    [LN]     $(2) $$@"
 	$(qexec)mkdir -p $$(dir $$@)
 	$(qexec)ln -sf $(2) $$@
@@ -318,14 +315,15 @@ endef
 
 $(eval $(call libvpx_symlink_template,\
     $(addprefix $(BUILD_PFX),$(notdir $(LIBVPX_SO_SYMLINKS))),\
+    $(BUILD_PFX)$(LIBVPX_SO),\
     $(BUILD_PFX)$(LIBVPX_SO)))
 $(eval $(call libvpx_symlink_template,\
     $(addprefix $(DIST_DIR)/,$(LIBVPX_SO_SYMLINKS)),\
-    $(LIBVPX_SO)))
+    $(LIBVPX_SO),\
+    $(DIST_DIR)/bin/$(LIBVPX_SO)))
 
 
-INSTALL-LIBS-$(CONFIG_SHARED) += $(LIBVPX_SO_SYMLINKS)
-INSTALL-LIBS-$(CONFIG_SHARED) += $(LIBSUBDIR)/$(LIBVPX_SO)
+INSTALL-LIBS-$(CONFIG_SHARED) += $(BINSUBDIR)/$(LIBVPX_SO)
 
 
 LIBS-yes += vpx.pc
@@ -487,7 +485,7 @@ OBJS-$(BUILD_LIBVPX) += $(LIBVPX_TEST_OBJS)
 BINS-yes += $(LIBVPX_TEST_BIN)
 
 CODEC_LIB=$(if $(CONFIG_DEBUG_LIBS),vpx_g,vpx)
-CODEC_LIB_SUF=$(if $(CONFIG_SHARED),.so,.a)
+CODEC_LIB_SUF=$(if $(CONFIG_SHARED),.dll,.a)
 TEST_LIBS := lib$(CODEC_LIB)$(CODEC_LIB_SUF) libgtest.a
 $(LIBVPX_TEST_BIN): $(TEST_LIBS)
 $(eval $(call linkerxx_template,$(LIBVPX_TEST_BIN), \
