From add0190038800e27451f85da191ba993dcd0e0ef Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 19 Apr 2015 21:48:21 +0900
Subject: [PATCH 4/6] configure: Search dirs under ${build_sysroot} instead of
 ${prefix}.

---
 configure | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/configure b/configure
index 6d7152e..47bbc33 100755
--- a/configure
+++ b/configure
@@ -7578,7 +7578,14 @@ case "$target" in
   x86_64-*mingw* | *-w64-mingw*)
   # MinGW-w64 does not use newlib, nor does it use winsup. It may,
   # however, use a symlink named 'mingw' in ${prefix} .
-    FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/mingw/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/mingw/include'
+    # Check whether --with-build-sysroot was given.
+    if test "${with_build_sysroot+set}" = set; then :
+    withval=$with_build_sysroot; if test x"$withval" != x ; then
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET" -L${withval}/${target}/lib -L${withval}/lib -isystem ${withval}/${target}/include -isystem ${withval}/include"
+    fi
+    else
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/include'
+    fi
     ;;
   *-mingw*)
   # MinGW can't be handled as Cygwin above since it does not use newlib.
-- 
2.5.1

