From 9adc737ae251824cd776ba6c774f7767a7769c61 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 19 Apr 2015 21:48:21 +0900
Subject: [PATCH 04/11] configure: Search dirs under ${build_sysroot} instead
 of ${prefix}.

---
 configure | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/configure b/configure
index a5f4fc5..3f80dfe 100755
--- a/configure
+++ b/configure
@@ -7619,7 +7619,14 @@ case "$target" in
   x86_64-*mingw* | *-w64-mingw*)
   # MinGW-w64 does not use newlib, nor does it use winsup. It may,
   # however, use a symlink named 'mingw' in ${prefix} .
-    FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/mingw/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/mingw/include'
+    # Check whether --with-build-sysroot was given.
+    if test "${with_build_sysroot+set}" = set; then :
+    withval=$with_build_sysroot; if test x"$withval" != x ; then
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET" -L${withval}/${target}/lib -L${withval}/lib -isystem ${withval}/${target}/include -isystem ${withval}/include"
+    fi
+    else
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/include'
+    fi
     ;;
   *-mingw*)
   # MinGW can't be handled as Cygwin above since it does not use newlib.
-- 
2.10.1

