--- binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/bfd/Makefile.am.orig	2014-08-01 21:54:27.917294900 +0400
+++ binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/bfd/Makefile.am	2014-08-01 23:41:39.388829543 +0400
@@ -833,7 +833,8 @@
 EXTRA_libbfd_la_SOURCES = $(CFILES)
 libbfd_la_DEPENDENCIES = $(OFILES) ofiles
 libbfd_la_LIBADD = `cat ofiles` @SHARED_LIBADD@ $(LIBDL)
-libbfd_la_LDFLAGS += -release `cat libtool-soversion` @SHARED_LDFLAGS@ -Wl,--whole-archive -Wl,$(LIBIBERTY) -Wl,$(LIBINTL)  -Wl,--no-whole-archive
+comma := ,
+libbfd_la_LDFLAGS += -release `cat libtool-soversion` @SHARED_LDFLAGS@ -Wl,--whole-archive -Wl,$(LIBIBERTY) $(addprefix -Wl$(comma),$(LIBINTL)) -Wl,--no-whole-archive
 
 # libtool will build .libs/libbfd.a.  We create libbfd.a in the build
 # directory so that we don't have to convert all the programs that use
--- binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/bfd/Makefile.in.orig	2014-08-01 21:54:27.917294900 +0400
+++ binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/bfd/Makefile.in	2014-08-01 23:42:11.628451359 +0400
@@ -341,8 +341,9 @@
 CSEARCH = -I. -I$(srcdir) -I$(INCDIR)
 SUBDIRS = doc po
 bfddocdir = doc
+comma := ,
 libbfd_la_LDFLAGS = $(am__append_1) -release `cat libtool-soversion` \
-	@SHARED_LDFLAGS@ $(am__empty) -no-undefined -Wl,--whole-archive -Wl,$(LIBIBERTY) -Wl,$(LIBINTL) -Wl,--no-whole-archive
+	@SHARED_LDFLAGS@ $(am__empty) -no-undefined -Wl,--whole-archive -Wl,$(LIBIBERTY) $(addprefix -Wl$(comma),$(LIBINTL)) -Wl,--no-whole-archive
 @INSTALL_LIBBFD_TRUE@bfdlib_LTLIBRARIES = libbfd.la
 @INSTALL_LIBBFD_FALSE@bfdinclude_HEADERS = $(am__append_2)
 @INSTALL_LIBBFD_TRUE@bfdinclude_HEADERS = $(BFD_H) \
--- binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/intl/localcharset.c.orig	2014-07-24 17:46:54.721150000 +0400
+++ binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/intl/localcharset.c	2014-08-01 23:46:23.527989681 +0400
@@ -23,6 +23,10 @@
 # include <config.h>
 #endif
 
+#ifndef STATIC
+#define STATIC static
+#endif
+
 /* Specification.  */
 #include "localcharset.h"
 
--- binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/intl/localcharset.h.orig	2014-07-24 17:46:54.721150000 +0400
+++ binutils-git-90debf20f4cc4ddd7cfb6356fe0d3876a18604a6/intl/localcharset.h	2014-08-01 23:48:01.816349439 +0400
@@ -31,7 +31,12 @@
    The result must not be freed; it is statically allocated.
    If the canonical name cannot be determined, the result is a non-canonical
    name.  */
-extern const char * locale_charset (void);
+#ifdef STATIC
+STATIC
+#else
+extern
+#endif
+const char * locale_charset (void);
 
 
 #ifdef __cplusplus
diff --git a/binutils/configure b/binutils/configure
index 4198e12..ec4f61f 100755
--- a/binutils/configure
+++ b/binutils/configure
@@ -13571,9 +13571,9 @@ fi
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -13601,9 +13601,9 @@ fi
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
diff --git a/intl/configure b/intl/configure
index 34da38b..4578bdb 100755
--- a/intl/configure
+++ b/intl/configure
@@ -5097,9 +5097,9 @@ fi
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -5127,9 +5127,9 @@ fi
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
@@ -5889,9 +5889,9 @@ fi
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -5919,9 +5919,9 @@ fi
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
diff --git a/opcodes/Makefile.am b/opcodes/Makefile.am
index 4acc436..46f9763 100644
--- a/opcodes/Makefile.am
+++ b/opcodes/Makefile.am
@@ -320,7 +320,7 @@ libopcodes_la_SOURCES =  dis-buf.c disassemble.c dis-init.c
 # hardcoded path to libbfd.so instead of relying on the entries in libbfd.la.
 libopcodes_la_DEPENDENCIES = $(OFILES) @SHARED_DEPENDENCIES@
 libopcodes_la_LIBADD = $(OFILES) @SHARED_LIBADD@
-libopcodes_la_LDFLAGS += -release `cat ../bfd/libtool-soversion` @SHARED_LDFLAGS@
+libopcodes_la_LDFLAGS += -release `cat ../bfd/libtool-soversion` @SHARED_LDFLAGS@ -Wl,$(LIBIBERTY) $(addprefix -Wl$(comma),$(LIBINTL))
 # Allow dependency tracking to work on all the source files.
 EXTRA_libopcodes_la_SOURCES = $(LIBOPCODES_CFILES)
 
diff --git a/opcodes/Makefile.in b/opcodes/Makefile.in
index 46ef017..8654aff 100644
--- a/opcodes/Makefile.in
+++ b/opcodes/Makefile.in
@@ -319,8 +319,9 @@ COMPILE_FOR_BUILD = $(CC_FOR_BUILD) $(INCLUDES) $(AM_CPPFLAGS) \
 LINK_FOR_BUILD = $(CC_FOR_BUILD) $(CFLAGS_FOR_BUILD) \
 	$(LDFLAGS_FOR_BUILD) -o $@
 
+comma := ,
 libopcodes_la_LDFLAGS = $(am__append_1) -release `cat \
-	../bfd/libtool-soversion` @SHARED_LDFLAGS@
+	../bfd/libtool-soversion` @SHARED_LDFLAGS@ -Wl,$(LIBIBERTY) $(addprefix -Wl$(comma),$(LIBINTL))
 @INSTALL_LIBBFD_TRUE@bfdlib_LTLIBRARIES = libopcodes.la
 @INSTALL_LIBBFD_TRUE@bfdinclude_DATA = $(INCDIR)/dis-asm.h
 @INSTALL_LIBBFD_FALSE@rpath_bfdlibdir = @bfdlibdir@
