--- gettext-0.19.2/gettext-runtime/intl/localcharset.c	2014-09-11 15:21:22.247551300 +0900
+++ gettext-0.19.2/gettext-runtime/intl/localcharset.c	2014-09-11 15:39:14.073457900 +0900
@@ -461,14 +461,7 @@
 
   static char buf[2 + 10 + 1];
 
-  /* The Windows API has a function returning the locale's codepage as a
-     number: GetACP().
-     When the output goes to a console window, it needs to be provided in
-     GetOEMCP() encoding if the console is using a raster font, or in
-     GetConsoleOutputCP() encoding if it is using a TrueType font.
-     But in GUI programs and for output sent to files and pipes, GetACP()
-     encoding is the best bet.  */
-  sprintf (buf, "CP%u", GetACP ());
+  sprintf (buf, "CP%u", GetConsoleOutputCP ());
   codeset = buf;
 
 #elif defined OS2
--- gettext-0.19.2/gettext-runtime/gnulib-lib/localcharset.c	2014-07-14 16:28:09.000000000 +0900
+++ gettext-0.19.2/gettext-runtime/gnulib-lib/localcharset.c	2014-09-11 15:42:09.016464100 +0900
@@ -461,14 +461,7 @@
 
   static char buf[2 + 10 + 1];
 
-  /* The Windows API has a function returning the locale's codepage as a
-     number: GetACP().
-     When the output goes to a console window, it needs to be provided in
-     GetOEMCP() encoding if the console is using a raster font, or in
-     GetConsoleOutputCP() encoding if it is using a TrueType font.
-     But in GUI programs and for output sent to files and pipes, GetACP()
-     encoding is the best bet.  */
-  sprintf (buf, "CP%u", GetACP ());
+  sprintf (buf, "CP%u", GetConsoleOutputCP ());
   codeset = buf;
 
 #elif defined OS2
