From 47f11888906fe9e111b648fce002129e91861e4d Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sat, 11 Apr 2015 11:48:43 +0900
Subject: [PATCH 2/6] Makefile: Setting PKG_CONFIG_PATH in Makefile is
 nonsense.

---
 Makefile | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index ff2915f..d6ee50b 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,6 @@ PREFIX ?= /usr/local
 BINDIR ?= $(PREFIX)/bin
 LIBDIR ?= $(PREFIX)/lib
 INCLUDEDIR ?= $(PREFIX)/include
-PKG_CONFIG_PATH ?= $(LIBDIR)/pkgconfig
 
 -include .config
 
@@ -129,8 +128,8 @@ dcadec.pc: dcadec.pc.in
 	sed 's,%PREFIX%,$(PREFIX),;s,%LIBDIR%,$(LIBDIR),;s,%INCLUDEDIR%,$(INCLUDEDIR),;s,%VERSION%,$(VERSION),' $< > $@
 
 install-lib: $(OUT_LIB) dcadec.pc
-	install -d -m 755 $(DESTDIR)$(LIBDIR) $(DESTDIR)$(PKG_CONFIG_PATH) $(DESTDIR)$(INCLUDEDIR)/libdcadec
+	install -d -m 755 $(DESTDIR)$(LIBDIR) $(DESTDIR)$(LIBDIR)/pkgconfig $(DESTDIR)$(INCLUDEDIR)/libdcadec
 	install -m 644 $(OUT_LIB) $(DESTDIR)$(LIBDIR)
 	install -m 644 $(addprefix $(SRC_DIR)/, $(INC_LIB)) $(DESTDIR)$(INCLUDEDIR)/libdcadec
-	install -m 644 dcadec.pc $(DESTDIR)$(PKG_CONFIG_PATH)
+	install -m 644 dcadec.pc $(DESTDIR)$(LIBDIR)/pkgconfig
 
-- 
2.3.7

