From 8ccf70976955ea6b3173af1781dd1d1ac0654ea0 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Wed, 29 Apr 2015 00:24:00 +0900
Subject: [PATCH 6/6] Makefile: Add API version suffix to DLL Name.

---
 Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3820a3f..7a86bef 100644
--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,8 @@ BINDIR ?= $(PREFIX)/bin
 LIBDIR ?= $(PREFIX)/lib
 INCLUDEDIR ?= $(PREFIX)/include
 
+API_VER=$(shell grep -e '\#define DCADEC_VERSION_MAJOR' libdcadec/dca_context.h | awk '{ printf $$3 }')
+
 -include .config
 
 ifdef CONFIG_NDEBUG
@@ -26,7 +28,7 @@ else
 endif
 
 ifdef CONFIG_SHARED
-    OUT_LIB ?= libdcadec/libdcadec$(DLLSUF)
+    OUT_LIB ?= libdcadec/libdcadec-$(API_VER)$(DLLSUF)
 else
     OUT_LIB ?= libdcadec/libdcadec$(LIBSUF)
 endif
-- 
2.3.7

