From 95c74c072ddc9be50df6b06823aa73879c928e04 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Wed, 29 Apr 2015 01:18:57 +0900
Subject: [PATCH 2/2] Add API version suffix to DLL Name.

---
 Makefile                   | 8 ++++++++
 build/platform-mingw_nt.mk | 2 ++
 2 files changed, 10 insertions(+)

diff --git a/Makefile b/Makefile
index d415cfc..0242897 100644
--- a/Makefile
+++ b/Makefile
@@ -216,7 +216,11 @@ $(LIBPREFIX)$(PROJECT_NAME).$(LIBSUFFIX): $(ENCODER_OBJS) $(DECODER_OBJS) $(PROC
 
 $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIXVER): $(ENCODER_OBJS) $(DECODER_OBJS) $(PROCESSING_OBJS) $(COMMON_OBJS)
 	$(QUIET)rm -f $@
+ifneq (mingw_nt, $(OS))
 	$(QUIET_CXX)$(CXX) $(SHARED) $(CXX_LINK_O) $+ $(LDFLAGS) $(SHLDFLAGS)
+else
+	$(QUIET_CXX)$(CXX) $(SHARED) -o $(LIBPREFIX)$(APIED_PROJECT_NAME).$(SHAREDLIBSUFFIXVER) $+ $(LDFLAGS) $(SHLDFLAGS)
+endif
 
 ifneq ($(SHAREDLIBSUFFIXVER),$(SHAREDLIBSUFFIX))
 $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIX): $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIXVER)
@@ -261,7 +265,11 @@ install-static: install-static-lib $(PROJECT_NAME)-static.pc
 
 install-shared: $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIX) install-headers $(PROJECT_NAME).pc
 	mkdir -p $(DESTDIR)$(SHAREDLIB_DIR)
+ifneq (mingw_nt, $(OS))
 	install -m 755 $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIXVER) $(DESTDIR)$(SHAREDLIB_DIR)
+else
+	install -m 755 $(LIBPREFIX)$(APIED_PROJECT_NAME).$(SHAREDLIBSUFFIXVER) $(DESTDIR)$(SHAREDLIB_DIR)
+endif
 	if [ "$(SHAREDLIBSUFFIXVER)" != "$(SHAREDLIBSUFFIX)" ]; then \
 		cp -a $(LIBPREFIX)$(PROJECT_NAME).$(SHAREDLIBSUFFIX) $(DESTDIR)$(SHAREDLIB_DIR); \
 	fi
diff --git a/build/platform-mingw_nt.mk b/build/platform-mingw_nt.mk
index f654661..d521f38 100644
--- a/build/platform-mingw_nt.mk
+++ b/build/platform-mingw_nt.mk
@@ -2,6 +2,8 @@ include $(SRC_PATH)build/x86-common.mk
 SHAREDLIB_DIR = $(PREFIX)/bin
 SHAREDLIBSUFFIX = dll
 SHAREDLIBSUFFIXVER=$(SHAREDLIBSUFFIX)
+API_VER=$(shell grep -e '\#define OPENH264_MAJOR' $(SRC_PATH)codec/api/svc/codec_ver.h | awk '{ printf $$3 }' | sed -e "s/^.*(\(.*\)).*$$/\1/")
+APIED_PROJECT_NAME=$(PROJECT_NAME)-$(API_VER)
 EXTRA_LIBRARY=$(LIBPREFIX)$(PROJECT_NAME).dll.a
 SHLDFLAGS = -Wl,--out-implib,$(EXTRA_LIBRARY) -Wl,--version-script,libopenh264.ver
 CFLAGS += -MMD -MP
-- 
2.5.0

