--- gcc-4.9.2/libgcc/gthr-posix.h.old	2014-10-28 16:14:25.340248300 +0900
+++ gcc-4.9.2/libgcc/gthr-posix.h	2014-10-28 16:15:07.806640900 +0900
@@ -82,7 +82,7 @@
 # define __GTHREAD_COND_INIT_FUNCTION __gthread_cond_init_function
 #endif
 
-#if SUPPORTS_WEAK && GTHREAD_USE_WEAK
+#if SUPPORTS_WEAK && GTHREAD_USE_WEAK && !defined(__MINGW32__)
 # ifndef __gthrw_pragma
 #  define __gthrw_pragma(pragma)
 # endif
