From 761d163f4841e20d1e528be041068854772b3240 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 12 Apr 2015 11:44:14 +0900
Subject: [PATCH 03/24] libiberty/Makefile.in: Don't install libiberty.

---
 libiberty/Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index e919a57..168227a 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -355,7 +355,7 @@ libiberty.html : $(srcdir)/libiberty.texi $(TEXISRC)
 @MAINT@	echo stamp > stamp-functions
 
 INSTALL_DEST = @INSTALL_DEST@
-install: install_to_$(INSTALL_DEST) install-subdir
+install: install-subdir
 install-strip: install
 
 .PHONY: install install-strip
-- 
2.8.1

