--- gcc-4.9.2/libgcc/gthr-posix.h.old	2014-10-28 16:32:18.199771800 +0900
+++ gcc-4.9.2/libgcc/gthr-posix.h	2014-10-28 16:32:47.173429000 +0900
@@ -34,6 +34,10 @@
 
 #include <pthread.h>
 
+#if defined (_WIN32) && !defined (__CYGWIN__)
+#define _GTHREAD_USE_MUTEX_INIT_FUNC 1
+#endif
+
 #if ((defined(_LIBOBJC) || defined(_LIBOBJC_WEAK)) \
      || !defined(_GTHREAD_USE_MUTEX_TIMEDLOCK))
 # include <unistd.h>
