--- gcc-4.9.2/libgomp/team.c.old	2014-01-03 07:25:54.000000000 +0900
+++ gcc-4.9.2/libgomp/team.c	2014-10-28 17:08:28.283072100 +0900
@@ -145,7 +145,7 @@
 
   size = sizeof (*team) + nthreads * (sizeof (team->ordered_release[0])
 				      + sizeof (team->implicit_task[0]));
-  team = gomp_malloc (size);
+  team = gomp_malloc_cleared (size);
 
   team->work_share_chunk = 8;
 #ifdef HAVE_SYNC_BUILTINS
@@ -196,7 +196,7 @@
 static struct gomp_thread_pool *gomp_new_thread_pool (void)
 {
   struct gomp_thread_pool *pool
-    = gomp_malloc (sizeof(struct gomp_thread_pool));
+    = gomp_malloc_cleared (sizeof(struct gomp_thread_pool));
   pool->threads = NULL;
   pool->threads_size = 0;
   pool->threads_used = 0;
@@ -939,7 +939,7 @@
 gomp_new_icv (void)
 {
   struct gomp_thread *thr = gomp_thread ();
-  struct gomp_task *task = gomp_malloc (sizeof (struct gomp_task));
+  struct gomp_task *task = gomp_malloc_cleared (sizeof (struct gomp_task));
   gomp_init_task (task, NULL, &gomp_global_icv);
   thr->task = task;
   pthread_setspecific (gomp_thread_destructor, thr);
--- gcc-4.9.2/libgomp/testsuite/config/default.exp.old	2014-01-03 07:25:54.000000000 +0900
+++ gcc-4.9.2/libgomp/testsuite/config/default.exp	2014-10-28 17:09:12.242586500 +0900
@@ -15,3 +15,7 @@
 # <http://www.gnu.org/licenses/>.
 
 load_lib "standard.exp"
+
+# Support for old dejagnu.  Must be loaded here, not in libstdc++.exp, to
+# make sure all existing procs are loaded when their presence is tested.
+load_file $srcdir/../../gcc/testsuite/lib/dejapatches.exp
--- gcc-4.9.2/libgomp/work.c.old	2014-01-03 07:25:54.000000000 +0900
+++ gcc-4.9.2/libgomp/work.c	2014-10-28 17:10:11.633983500 +0900
@@ -75,7 +75,7 @@
 #endif
 
   team->work_share_chunk *= 2;
-  ws = gomp_malloc (team->work_share_chunk * sizeof (struct gomp_work_share));
+  ws = gomp_malloc_cleared (team->work_share_chunk * sizeof (struct gomp_work_share));
   ws->next_alloc = team->work_shares[0].next_alloc;
   team->work_shares[0].next_alloc = ws;
   team->work_share_list_alloc = &ws[1];
@@ -174,7 +174,7 @@
   /* Work sharing constructs can be orphaned.  */
   if (team == NULL)
     {
-      ws = gomp_malloc (sizeof (*ws));
+      ws = gomp_malloc_cleared (sizeof (*ws));
       gomp_init_work_share (ws, ordered, 1);
       thr->ts.work_share = ws;
       return ws;
