--- gcc-4.9.2/gcc/diagnostic-color.c.old	2014-01-03 07:23:26.000000000 +0900
+++ gcc-4.9.2/gcc/diagnostic-color.c	2014-10-28 17:17:20.815531200 +0900
@@ -263,20 +263,17 @@
       return true;
 }
 
-#if defined(_WIN32)
-bool
-colorize_init (diagnostic_color_rule_t)
-{
-  return false;
-}
-#else
-
 /* Return true if we should use color when in auto mode, false otherwise. */
 static bool
 should_colorize (void)
 {
   char const *t = getenv ("TERM");
+#ifdef __MINGW32__
+  char const *mintty = getenv ("MINTTY");
+  return t && strcmp (t, "dumb") != 0 && (_isatty (STDERR_FILENO) || (mintty && strcmp (mintty, "NO_FILE") == 0));
+#else
   return t && strcmp (t, "dumb") != 0 && isatty (STDERR_FILENO);
+#endif
 }
 
 
@@ -298,4 +295,3 @@
       gcc_unreachable ();
     }
 }
-#endif
