From 410270082bb9e70b16bf4c856a80a58fe772d945 Mon Sep 17 00:00:00 2001
From: Yuta Nakai <nak5124@live.jp>
Date: Sun, 24 Apr 2016 01:28:05 +0900
Subject: [PATCH 11/24] gcc: Add prefix/bindir to exec_prefix.

---
 gcc/gcc.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/gcc/gcc.c b/gcc/gcc.c
index f7434a1..9e1d99c 100644
--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -4118,7 +4118,7 @@ process_command (unsigned int decoded_options_count,
 {
   const char *temp;
   char *temp1;
-  char *tooldir_prefix, *tooldir_prefix2;
+  char *tooldir_prefix, *tooldir_prefix2, *bindir_prefix, *bindir_prefix2;
   char *(*get_relative_prefix) (const char *, const char *,
 				const char *) = NULL;
   struct cl_option_handlers handlers;
@@ -4487,23 +4487,34 @@ process_command (unsigned int decoded_options_count,
     }
 
   gcc_assert (!IS_ABSOLUTE_PATH (tooldir_base_prefix));
+  bindir_prefix2 = concat (tooldir_base_prefix, "bin",
+			    dir_separator_str, NULL);
   tooldir_prefix2 = concat (tooldir_base_prefix, spec_machine,
 			    dir_separator_str, NULL);
 
   /* Look for tools relative to the location from which the driver is
      running, or, if that is not available, the configured prefix.  */
+  bindir_prefix
+    = concat (gcc_exec_prefix ? gcc_exec_prefix : standard_exec_prefix,
+	      spec_machine, dir_separator_str,
+	      spec_version, dir_separator_str, bindir_prefix2, NULL);
   tooldir_prefix
     = concat (gcc_exec_prefix ? gcc_exec_prefix : standard_exec_prefix,
 	      spec_host_machine, dir_separator_str, spec_version,
 	      accel_dir_suffix, dir_separator_str, tooldir_prefix2, NULL);
+  free (bindir_prefix2);
   free (tooldir_prefix2);
 
   add_prefix (&exec_prefixes,
+	      bindir_prefix,
+	      "BINUTILS", PREFIX_PRIORITY_LAST, 0, 0);
+  add_prefix (&exec_prefixes,
 	      concat (tooldir_prefix, "bin", dir_separator_str, NULL),
 	      "BINUTILS", PREFIX_PRIORITY_LAST, 0, 0);
   add_prefix (&startfile_prefixes,
 	      concat (tooldir_prefix, "lib", dir_separator_str, NULL),
 	      "BINUTILS", PREFIX_PRIORITY_LAST, 0, 1);
+  free (bindir_prefix);
   free (tooldir_prefix);
 
 #if defined(TARGET_SYSTEM_ROOT_RELOCATABLE) && !defined(VMS)
-- 
2.8.1

