--- gcc-4.9.2/configure.old	2014-07-27 00:18:04.000000000 +0900
+++ gcc-4.9.2/configure	2014-10-28 17:22:42.027947600 +0900
@@ -7041,6 +7041,12 @@
 # using) don't handle "\$" correctly, so don't use it here.
 tooldir='${exec_prefix}'/${target_noncanonical}
 build_tooldir=${tooldir}
+# Check whether --with-build-sysroot was given.
+if test "${with_build_sysroot+set}" = set; then :
+  withval=$with_build_sysroot; if test x"$withval" != x ; then
+  build_tooldir=${withval}/${target_noncanonical}
+  fi
+fi
 
 # Create a .gdbinit file which runs the one in srcdir
 # and tells GDB to look there for source files.
@@ -7648,7 +7654,14 @@
   x86_64-*mingw* | *-w64-mingw*)
   # MinGW-w64 does not use newlib, nor does it use winsup. It may,
   # however, use a symlink named 'mingw' in ${prefix} .
-    FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/mingw/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/mingw/include'
+    # Check whether --with-build-sysroot was given.
+    if test "${with_build_sysroot+set}" = set; then :
+    withval=$with_build_sysroot; if test x"$withval" != x ; then
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET" -L${withval}/${target}/lib -L${withval}/lib -isystem ${withval}/${target}/include -isystem ${withval}/include"
+    fi
+    else
+        FLAGS_FOR_TARGET=$FLAGS_FOR_TARGET' -L${prefix}/${target}/lib -L${prefix}/lib -isystem ${prefix}/${target}/include -isystem ${prefix}/include'
+    fi
     ;;
   *-mingw*)
   # MinGW can't be handled as Cygwin above since it does not use newlib.
--- gcc-4.9.2/gcc/configure.old	2014-10-10 21:51:28.000000000 +0900
+++ gcc-4.9.2/gcc/configure	2014-10-28 17:24:01.617499900 +0900
@@ -670,6 +670,7 @@
 srcdir
 subdirs
 dollar
+build_tooldir
 gcc_tooldir
 enable_lto
 DO_LINK_MUTEX
@@ -7158,9 +7159,11 @@
 if test "${with_build_sysroot+set}" = set; then :
   withval=$with_build_sysroot; if test x"$withval" != x ; then
      SYSROOT_CFLAGS_FOR_TARGET="--sysroot=$withval"
+     build_tooldir=${withval}/${target_noncanonical}
    fi
 else
   SYSROOT_CFLAGS_FOR_TARGET=
+  build_tooldir='${exec_prefix}'/${target_noncanonical}
 fi
 
 
--- gcc-4.9.2/gcc/Makefile.in.old	2014-10-16 22:50:42.000000000 +0900
+++ gcc-4.9.2/gcc/Makefile.in	2014-10-28 17:24:55.000553200 +0900
@@ -632,7 +632,7 @@
 # Used in install-cross.
 gcc_tooldir = @gcc_tooldir@
 # Since gcc_tooldir does not exist at build-time, use -B$(build_tooldir)/bin/
-build_tooldir = $(exec_prefix)/$(target_noncanonical)
+build_tooldir = @build_tooldir@
 # Directory in which the compiler finds target-independent g++ includes.
 gcc_gxx_include_dir = @gcc_gxx_include_dir@
 gcc_gxx_include_dir_add_sysroot = @gcc_gxx_include_dir_add_sysroot@
