--- gcc-4.9.1/gcc/configure.old	2014-07-24 12:33:43.215340900 +0900
+++ gcc-4.9.1/gcc/configure	2014-07-24 11:59:16.432238600 +0900
@@ -3329,7 +3329,8 @@
 # This logic must match libstdc++-v3/acinclude.m4:GLIBCXX_EXPORT_INSTALL_INFO.
 if test x${gcc_gxx_include_dir} = x; then
   if test x${enable_version_specific_runtime_libs} = xyes; then
-    gcc_gxx_include_dir='${libsubdir}/include/c++'
+    # gcc_gxx_include_dir='${libsubdir}/include/c++'
+    gcc_gxx_include_dir='${prefix}/include/c++/$(version)'
   else
     libstdcxx_incdir='include/c++/$(version)'
     if test x$host != x$target; then
--- gcc-4.9.1/libstdc++-v3/configure.old	2014-07-24 12:34:32.167057000 +0900
+++ gcc-4.9.1/libstdc++-v3/configure	2014-07-24 12:00:14.198319900 +0900
@@ -79462,7 +79462,8 @@
     # is selected.  FIXME: these variables are misnamed, there are
     # no executables installed in _toolexecdir or _toolexeclibdir.
     if test x"$gxx_include_dir" = x"no"; then
-      gxx_include_dir='${libdir}/gcc/${host_alias}/${gcc_version}/include/c++'
+      # gxx_include_dir='${libdir}/gcc/${host_alias}/${gcc_version}/include/c++'
+      gxx_include_dir='${prefix}/include/c++/${gcc_version}'
     fi
     glibcxx_toolexecdir='${libdir}/gcc/${host_alias}'
     glibcxx_toolexeclibdir='${toolexecdir}/${gcc_version}$(MULTISUBDIR)'
