--- gcc-4.9.2/gcc/configure.old	2014-10-28 17:24:01.617499900 +0900
+++ gcc-4.9.2/gcc/configure	2014-10-28 17:49:05.424557600 +0900
@@ -9892,9 +9892,9 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -9922,9 +9922,9 @@
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
--- gcc-4.9.2/intl/configure.old	2010-09-28 05:19:41.000000000 +0900
+++ gcc-4.9.2/intl/configure	2014-10-28 17:49:05.438558400 +0900
@@ -5097,9 +5097,9 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -5127,9 +5127,9 @@
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
@@ -5889,9 +5889,9 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -5919,9 +5919,9 @@
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
--- gcc-4.9.2/libcpp/configure.old	2014-10-23 20:12:54.000000000 +0900
+++ gcc-4.9.2/libcpp/configure	2014-10-28 17:49:05.451559100 +0900
@@ -6605,9 +6605,9 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -6635,9 +6635,9 @@
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
--- gcc-4.9.2/libstdc++-v3/configure.old	2014-03-21 03:57:07.000000000 +0900
+++ gcc-4.9.2/libstdc++-v3/configure	2014-10-28 17:49:05.503562100 +0900
@@ -27275,9 +27275,9 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              if test -f "$additional_libdir/lib$name.dll.a"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
+                found_a="$additional_libdir/lib$name.dll.a"
                 if test -f "$additional_libdir/lib$name.la"; then
                   found_la="$additional_libdir/lib$name.la"
                 fi
@@ -27305,9 +27305,9 @@
                       found_la="$dir/lib$name.la"
                     fi
                   else
-                    if test -f "$dir/lib$name.$libext"; then
+                    if test -f "$dir/lib$name.dll.a"; then
                       found_dir="$dir"
-                      found_a="$dir/lib$name.$libext"
+                      found_a="$dir/lib$name.dll.a"
                       if test -f "$dir/lib$name.la"; then
                         found_la="$dir/lib$name.la"
                       fi
