diff -r a01d69a95cbd gcc/common.opt
--- a/gcc/common.opt	Wed Dec 26 19:22:17 2012 -0700
+++ b/gcc/common.opt	Sun Dec 30 13:56:06 2012 -0700
@@ -2513,6 +2513,9 @@
 static-libstdc++
 Driver
 
+shared-libstdc++
+Driver
+
 static-libgo
 Driver
 ; Documented for Go, but always accepted by driver.
diff -r a01d69a95cbd gcc/cp/g++spec.c
--- a/gcc/cp/g++spec.c	Wed Dec 26 19:22:17 2012 -0700
+++ b/gcc/cp/g++spec.c	Sun Dec 30 13:56:06 2012 -0700
@@ -64,8 +64,8 @@
   /* What do with libstdc++:
      -1 means we should not link in libstdc++
      0  means we should link in libstdc++ if it is needed
-     1  means libstdc++ is needed and should be linked in.
-     2  means libstdc++ is needed and should be linked statically.  */
+     1  means libstdc++ is needed and should be linked in (statically).
+     2  means libstdc++ is needed and should be linked shared.  */
   int library = 0;
 
   /* The number of arguments being added to what's in argv, other than
@@ -98,7 +98,7 @@
   int static_link = 0;
 
   /* True if we should add -shared-libgcc to the command-line.  */
-  int shared_libgcc = 1;
+  int shared_libgcc = 0;
 
   /* The total number of arguments with the new stuff.  */
   unsigned int argc;
@@ -188,6 +188,11 @@
 	  break;
 
 	case OPT_static_libstdc__:
+	  library = library >= 0 ? 1 : library;
+	  args[i] |= SKIPOPT;
+	  break;
+
+	case OPT_shared_libstdc__:
 	  library = library >= 0 ? 2 : library;
 	  args[i] |= SKIPOPT;
 	  break;
@@ -314,7 +319,7 @@
   if (library > 0)
     {
 #ifdef HAVE_LD_STATIC_DYNAMIC
-      if (library > 1 && !static_link)
+      if (library == 1 && !static_link)
 	{
 	  generate_option (OPT_Wl_, LD_STATIC_OPTION, 1, CL_DRIVER,
 			   &new_decoded_options[j]);
@@ -327,7 +332,7 @@
       added_libraries++;
       j++;
       /* Add target-dependent static library, if necessary.  */
-      if ((static_link || library > 1) && LIBSTDCXX_STATIC != NULL)
+      if ((static_link || library == 1) && LIBSTDCXX_STATIC != NULL)
 	{
 	  generate_option (OPT_l, LIBSTDCXX_STATIC, 1,
 			   CL_DRIVER, &new_decoded_options[j]);
@@ -335,7 +340,7 @@
 	  j++;
 	}
 #ifdef HAVE_LD_STATIC_DYNAMIC
-      if (library > 1 && !static_link)
+      if (library == 1 && !static_link)
 	{
 	  generate_option (OPT_Wl_, LD_DYNAMIC_OPTION, 1, CL_DRIVER,
 			   &new_decoded_options[j]);
diff -r a01d69a95cbd gcc/gcc.c
--- a/gcc/gcc.c	Wed Dec 26 19:22:17 2012 -0700
+++ b/gcc/gcc.c	Sun Dec 30 13:56:06 2012 -0700
@@ -3585,9 +3585,10 @@
     case OPT_shared_libgcc:
     case OPT_static_libgfortran:
     case OPT_static_libstdc__:
+    case OPT_shared_libstdc__:
       /* These are always valid, since gcc.c itself understands the
 	 first two, gfortranspec.c understands -static-libgfortran and
-	 g++spec.c understands -static-libstdc++ */
+	 g++spec.c understands -static/shared-libstdc++ */
       validated = true;
       break;
 
diff -r e65b3ea63c66 gcc/config/i386/mingw-w64.h
--- a/gcc/config/i386/mingw-w64.h	Mon Sep 23 10:33:17 2013 -0600
+++ b/gcc/config/i386/mingw-w64.h	Sun Sep 29 07:45:07 2013 -0600
@@ -77,7 +77,7 @@
 #endif
 
 #undef LINK_SPEC
-#define LINK_SPEC SUB_LINK_SPEC " %{mwindows:--subsystem windows} \
+#define LINK_SPEC SUB_LINK_SPEC " --exclude-libs=libpthread.a %{mwindows:--subsystem windows} \
   %{mconsole:--subsystem console} \
   %{shared: %{mdll: %eshared and mdll are not compatible}} \
   %{shared: --shared} %{mdll:--dll} \
diff -r e65b3ea63c66 gcc/config/i386/mingw32.h
--- a/gcc/config/i386/mingw32.h	Mon Sep 23 10:33:17 2013 -0600
+++ b/gcc/config/i386/mingw32.h	Sun Sep 29 07:45:07 2013 -0600
@@ -110,7 +110,7 @@
 #define SUBTARGET_EXTRA_SPECS						\
   { "shared_libgcc_undefs", SHARED_LIBGCC_UNDEFS_SPEC }
 
-#define LINK_SPEC "%{mwindows:--subsystem windows} \
+#define LINK_SPEC "--exclude-libs=libpthread.a %{mwindows:--subsystem windows} \
   %{mconsole:--subsystem console} \
   %{shared: %{mdll: %eshared and mdll are not compatible}} \
   %{shared: --shared} %{mdll:--dll} \
