--- gcc-4.9.1/gcc/diagnostic-color.c.org	2014-09-11 16:01:36.439236900 +0900
+++ gcc-4.9.1/gcc/diagnostic-color.c	2014-09-11 02:12:14.227760800 +0900
@@ -268,7 +268,12 @@
 should_colorize (void)
 {
   char const *t = getenv ("TERM");
+#if defined(_WIN32) && defined(__MINGW32__)
+  char const *mintty = getenv ("MINTTY");
+  return t && strcmp (t, "dumb") != 0 && (_isatty (STDERR_FILENO) || (mintty && strcmp (mintty, "NO_FILE") == 0));
+#else
   return t && strcmp (t, "dumb") != 0 && isatty (STDERR_FILENO);
+#endif
 }
 
 
