--- gcc-4.9.2/gcc/config/i386/mingw32.h.old	2014-10-28 17:44:41.248402700 +0900
+++ gcc-4.9.2/gcc/config/i386/mingw32.h	2014-10-28 18:05:27.772401500 +0900
@@ -77,6 +77,22 @@
 #define SUB_LINK_ENTRY SUB_LINK_ENTRY32
 #endif
 
+#undef SUB_PIE_ENTRY32
+#undef SUB_PIE_ENTRY64
+#define SUB_PIE_ENTRY32 "-e _mainCRTStartup"
+#if defined(USE_MINGW64_LEADING_UNDERSCORES)
+#define SUB_PIE_ENTRY64 "-e _mainCRTStartup"
+#else
+#define SUB_PIE_ENTRY64 "-e mainCRTStartup"
+#endif
+
+#undef SUB_PIE_ENTRY
+#if TARGET_64BIT_DEFAULT
+#define SUB_PIE_ENTRY SUB_PIE_ENTRY64
+#else
+#define SUB_PIE_ENTRY SUB_PIE_ENTRY32
+#endif
+
 #undef NATIVE_SYSTEM_HEADER_COMPONENT
 #define NATIVE_SYSTEM_HEADER_COMPONENT "MINGW"
 
@@ -115,9 +131,10 @@
   %{shared: %{mdll: %eshared and mdll are not compatible}} \
   %{shared: --shared} %{mdll:--dll} \
   %{static:-Bstatic} %{!static:-Bdynamic} \
-  %{shared|mdll: " SUB_LINK_ENTRY " --enable-auto-image-base} \
+  --nxcompat --dynamicbase \
+  %{shared|mdll: " SUB_LINK_ENTRY " --disable-auto-image-base} \
   %(shared_libgcc_undefs) \
-  %{!shared: %{!mdll: --large-address-aware}}"
+  %{!shared: %{!mdll: --large-address-aware --pic-executable " SUB_PIE_ENTRY "}}"
 
 /* Include in the mingw32 libraries with libgcc */
 #ifdef ENABLE_SHARED_LIBGCC
--- gcc-4.9.2/gcc/config/i386/mingw-w64.h.old	2014-10-28 17:45:15.407356500 +0900
+++ gcc-4.9.2/gcc/config/i386/mingw-w64.h	2014-10-28 18:05:27.774401600 +0900
@@ -64,10 +64,21 @@
 #define SUB_LINK_ENTRY64 "-e DllMainCRTStartup"
 #endif
 
+#undef SUB_PIE_ENTRY32
+#undef SUB_PIE_ENTRY64
+#define SUB_PIE_ENTRY32 "-e _mainCRTStartup"
+#if defined(USE_MINGW64_LEADING_UNDERSCORES)
+#define SUB_PIE_ENTRY64 "-e _mainCRTStartup"
+#else
+#define SUB_PIE_ENTRY64 "-e mainCRTStartup"
+#endif
+
 #undef SUB_LINK_SPEC
 #undef SUB_LINK_ENTRY
+#undef SUB_PIE_ENTRY
 #define SUB_LINK_SPEC "%{" SPEC_64 ":-m i386pep} %{" SPEC_32 ":-m i386pe}"
 #define SUB_LINK_ENTRY "%{" SPEC_64 ":" SUB_LINK_ENTRY64 "} %{" SPEC_32 ":" SUB_LINK_ENTRY32 "}"
+#define SUB_PIE_ENTRY "%{" SPEC_64 ":" SUB_PIE_ENTRY64 "} %{" SPEC_32 ":" SUB_PIE_ENTRY32 "}"
 
 #undef MULTILIB_DEFAULTS
 #if TARGET_64BIT_DEFAULT
@@ -82,6 +93,8 @@
   %{shared: %{mdll: %eshared and mdll are not compatible}} \
   %{shared: --shared} %{mdll:--dll} \
   %{static:-Bstatic} %{!static:-Bdynamic} \
-  %{shared|mdll: " SUB_LINK_ENTRY " --enable-auto-image-base} \
+  --nxcompat --dynamicbase %{" SPEC_64 ": --high-entropy-va} \
+  %{shared|mdll: " SUB_LINK_ENTRY " %{" SPEC_64 ": --image-base 0x180000000} --disable-auto-image-base} \
   %(shared_libgcc_undefs) \
-  %{!shared: %{!mdll: %{" SPEC_32 ": --large-address-aware}}}"
+  %{!shared: %{!mdll: %{" SPEC_32 ": --large-address-aware} --pic-executable " SUB_PIE_ENTRY " \
+  %{" SPEC_64 ": --image-base 0x140000000}}}"
