--- gcc-4.9.1/gcc/intl.c.org	2014-09-11 15:47:29.230779300 +0900
+++ gcc-4.9.1/gcc/intl.c	2014-09-11 04:12:29.911343900 +0900
@@ -33,10 +33,10 @@
 const char *close_quote = "'";
 
 /* The name of the locale encoding.  */
-const char *locale_encoding = NULL;
+const char *locale_encoding = "utf-8";
 
 /* Whether the locale is using UTF-8.  */
-bool locale_utf8 = false;
+bool locale_utf8 = true;
 
 #ifdef ENABLE_NLS
 
@@ -78,13 +78,11 @@
 	 U+2018 and U+2019; but otherwise use "'" instead of "`" as
 	 opening quote.  */
       open_quote = "'";
-#if defined HAVE_LANGINFO_CODESET
       if (locale_utf8)
 	{
 	  open_quote = "\xe2\x80\x98";
 	  close_quote = "\xe2\x80\x99";
 	}
-#endif
     }
 }
 
--- gcc-4.9.1/gcc/pretty-print.c.org	2014-09-11 15:59:16.491232300 +0900
+++ gcc-4.9.1/gcc/pretty-print.c	2014-09-11 04:12:26.559918400 +0900
@@ -1073,7 +1073,7 @@
 
   /* Otherwise IDENT is converted to the locale character set if
      possible.  */
-#if defined ENABLE_NLS && defined HAVE_LANGINFO_CODESET && HAVE_ICONV
+#if defined ENABLE_NLS && HAVE_ICONV
   if (locale_encoding != NULL)
     {
       iconv_t cd = iconv_open (locale_encoding, "UTF-8");
