From 4b9eca200aba74c32683b17fa341d81ec3002dc1 Mon Sep 17 00:00:00 2001
From: Yuta Nakai <nak5124@live.jp>
Date: Sat, 24 Dec 2016 06:08:31 +0900
Subject: [PATCH 19/26] gcc/diagnostic-color.c: Enable color diagnostic on
 Windows.

---
 gcc/diagnostic-color.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/gcc/diagnostic-color.c b/gcc/diagnostic-color.c
index f76c87b..71d1326 100644
--- a/gcc/diagnostic-color.c
+++ b/gcc/diagnostic-color.c
@@ -264,14 +264,6 @@ parse_gcc_colors (void)
       return true;
 }
 
-#if defined(_WIN32)
-bool
-colorize_init (diagnostic_color_rule_t)
-{
-  return false;
-}
-#else
-
 /* Return true if we should use color when in auto mode, false otherwise. */
 static bool
 should_colorize (void)
@@ -299,4 +291,3 @@ colorize_init (diagnostic_color_rule_t rule)
       gcc_unreachable ();
     }
 }
-#endif
-- 
2.10.2

