From 6c113a878749b8ca520bf58e8c904238dc70a363 Mon Sep 17 00:00:00 2001
From: Yuta Nakai <nak5124@live.jp>
Date: Sun, 20 Mar 2016 15:53:32 +0900
Subject: [PATCH 19/23] MinGW-w64: Define __MSVCRT_VERSION__.

---
 gcc/config/i386/mingw-w64.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/gcc/config/i386/mingw-w64.h b/gcc/config/i386/mingw-w64.h
index ebeeddc..59ef622 100644
--- a/gcc/config/i386/mingw-w64.h
+++ b/gcc/config/i386/mingw-w64.h
@@ -24,6 +24,13 @@ along with GCC; see the file COPYING3.  If not see
 #undef CPP_SPEC
 #define CPP_SPEC "%{posix:-D_POSIX_SOURCE} %{mthreads:-D_MT} " \
 		 "%{municode:-DUNICODE} " \
+		 "%{mcrtdll=msvcrt|!mcrtdll=*:-D__MSVCRT_VERSION__=0x0700} " \
+		 "%{mcrtdll=msvcr80:-D__MSVCRT_VERSION__=0x0800} " \
+		 "%{mcrtdll=msvcr90:-D__MSVCRT_VERSION__=0x0900} " \
+		 "%{mcrtdll=msvcr100:-D__MSVCRT_VERSION__=0x1000} " \
+		 "%{mcrtdll=msvcr110:-D__MSVCRT_VERSION__=0x1100} " \
+		 "%{mcrtdll=msvcr120:-D__MSVCRT_VERSION__=0x1200} " \
+		 "%{mcrtdll=ucrtbase:-D__MSVCRT_VERSION__=0x1400 -D__UCRT__} " \
 		 PTHREAD_CPP_SPEC
 
 #undef STARTFILE_SPEC
-- 
2.8.1

