From 35896d05330901737eeeb5e314a0d9e3ef526d7b Mon Sep 17 00:00:00 2001
From: Yuta Nakai <nak5124@live.jp>
Date: Tue, 3 May 2016 00:31:25 +0900
Subject: [PATCH 24/24] MinGW-w64: Hack! Do not pass -iprefix internally.

---
 gcc/gcc.c | 20 +++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/gcc/gcc.c b/gcc/gcc.c
index 9e1d99c..be2a0f1 100644
--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -4574,10 +4574,10 @@ process_command (unsigned int decoded_options_count,
 
   if (print_version)
     n_varsafe_options++;
-  
+
   if (print_help_list)
     n_varsafe_options++;
-  
+
   spec_undefvar_allowed = (n_varsafe_options == decoded_options_count - 1);
 
   alloc_switch ();
@@ -5521,6 +5521,7 @@ do_spec_1 (const char *spec, int inswitch, const char *soft_matched_part)
 		  do_spec_1 (" ", 0, NULL);
 		}
 
+#ifndef __MINGW32__
 	      if (gcc_exec_prefix)
 		{
 		  do_spec_1 ("-iprefix", 1, NULL);
@@ -5529,6 +5530,7 @@ do_spec_1 (const char *spec, int inswitch, const char *soft_matched_part)
 		  do_spec_1 (gcc_exec_prefix, 1, NULL);
 		  do_spec_1 (" ", 0, NULL);
 		}
+#endif
 
 	      if (target_system_root_changed ||
 		  (target_system_root && target_sysroot_hdrs_suffix))
@@ -5853,7 +5855,7 @@ do_spec_1 (const char *spec, int inswitch, const char *soft_matched_part)
 		     "%{foo=*:bar%*}%{foo=*:one%*two}"
 
 		   matches -foo=hello then it will produce:
-		   
+
 		     barhello onehellotwo
 		*/
 		if (*p == 0 || *p == '}')
@@ -9528,9 +9530,9 @@ static unsigned HOST_WIDE_INT
 get_random_number (void)
 {
   unsigned HOST_WIDE_INT ret = 0;
-  int fd; 
+  int fd;
 
-  fd = open ("/dev/urandom", O_RDONLY); 
+  fd = open ("/dev/urandom", O_RDONLY);
   if (fd >= 0)
     {
       read (fd, &ret, sizeof (HOST_WIDE_INT));
@@ -9811,16 +9813,16 @@ greater_than_spec_func (int argc, const char **argv)
   return NULL;
 }
 
-/* Insert backslash before spaces in ORIG (usually a file path), to 
+/* Insert backslash before spaces in ORIG (usually a file path), to
    avoid being broken by spec parser.
 
    This function is needed as do_spec_1 treats white space (' ' and '\t')
    as the end of an argument. But in case of -plugin /usr/gcc install/xxx.so,
    the file name should be treated as a single argument rather than being
-   broken into multiple. Solution is to insert '\\' before the space in a 
+   broken into multiple. Solution is to insert '\\' before the space in a
    file name.
-   
-   This function converts and only converts all occurrence of ' ' 
+
+   This function converts and only converts all occurrence of ' '
    to '\\' + ' ' and '\t' to '\\' + '\t'.  For example:
    "a b"  -> "a\\ b"
    "a  b" -> "a\\ \\ b"
-- 
2.8.1

