diff --git a/demuxer/Demuxers/LAVFDemuxer.cpp b/demuxer/Demuxers/LAVFDemuxer.cpp
index 5330d7d..5c4765e 100644
--- a/demuxer/Demuxers/LAVFDemuxer.cpp
+++ b/demuxer/Demuxers/LAVFDemuxer.cpp
@@ -1424,6 +1424,7 @@ static int audio_codec_priority(AVCodecContext *codec)
 {
   int priority = 0;
   switch(codec->codec_id) {
+  case AV_CODEC_ID_ALAC:
   case AV_CODEC_ID_FLAC:
   case AV_CODEC_ID_TRUEHD:
   case AV_CODEC_ID_MLP:
