diff --git a/build.bat b/build.bat
index 88d60b5..d3e1847 100644
--- a/build.bat
+++ b/build.bat
@@ -2,10 +2,18 @@
 
 call "%VS120COMNTOOLS%vsvars32.bat"
 
-sh build_ffmpeg.sh x86
+REM sh build_ffmpeg.sh x86
 devenv LAVFilters.sln /Rebuild "Release|Win32"
 
-sh build_ffmpeg.sh x64
-devenv LAVFilters.sln /Rebuild "Release|x64"
+REM sh build_ffmpeg.sh x64
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+REM devenv LAVFilters.sln /Build "Release|x64"
+
+ISCC LAVFilters.iss
 
 PAUSE
diff --git a/build_ffmpeg.sh b/build_ffmpeg.sh
index 569e75f..06c7a9c 100644
--- a/build_ffmpeg.sh
+++ b/build_ffmpeg.sh
@@ -69,6 +69,8 @@ configure() (
     --enable-protocol=mmst          \
     --enable-protocol=rtp           \
     --enable-protocol=http          \
+    --enable-protocol=rtmp          \
+    --enable-protocol=rtmpt         \
     --disable-muxers                \
     --enable-muxer=spdif            \
     --disable-hwaccels              \
@@ -76,6 +78,8 @@ configure() (
     --enable-hwaccel=vc1_dxva2      \
     --enable-hwaccel=wmv3_dxva2     \
     --enable-hwaccel=mpeg2_dxva2    \
+    --disable-decoder=amrnb         \
+    --disable-decoder=amrwb         \
     --enable-libspeex               \
     --enable-libopencore-amrnb      \
     --enable-libopencore-amrwb      \
@@ -93,10 +98,9 @@ configure() (
     --build-suffix=-lav             \
     --arch=${arch}"
 
-  EXTRA_CFLAGS="-D_WIN32_WINNT=0x0502 -DWINVER=0x0502 -I../thirdparty/include"
+  EXTRA_CFLAGS="-D_WIN32_WINNT=0x0601 -DWINVER=0x0601 -I../thirdparty/include"
   EXTRA_LDFLAGS=""
   if [ "${arch}" == "x86_64" ]; then
-    OPTIONS="${OPTIONS} --enable-cross-compile --cross-prefix=x86_64-w64-mingw32- --target-os=mingw32"
     EXTRA_LDFLAGS="${EXTRA_LDFLAGS} -L../thirdparty/lib64"
   else
     OPTIONS="${OPTIONS} --cpu=i686"
