diff --git a/decoder/LAVVideo/decoders/avcodec.cpp b/decoder/LAVVideo/decoders/avcodec.cpp
index 5f31709..b08b6e4 100644
--- a/decoder/LAVVideo/decoders/avcodec.cpp
+++ b/decoder/LAVVideo/decoders/avcodec.cpp
@@ -373,7 +373,7 @@ STDMETHODIMP CDecAvcodec::InitDecoder(AVCodecID codec, const CMediaType *pmt)
   }
   m_pAVCtx->thread_count = max(1, min(thread_count, AVCODEC_MAX_THREADS));
 
-  if (dwDecFlags & LAV_VIDEO_DEC_FLAG_NO_MT || codec == AV_CODEC_ID_MPEG4) {
+  if (dwDecFlags & LAV_VIDEO_DEC_FLAG_NO_MT) {
     m_pAVCtx->thread_count = 1;
   }
 
