diff --git a/version.sh b/version.sh
index 79dd8ff..f30f461 100755
--- a/version.sh
+++ b/version.sh
@@ -18,9 +18,6 @@ if [ $LOCALVER \> 1 ] ; then
     if [ $VER_DIFF != 0 ] ; then
         VER="$VER+$VER_DIFF"
     fi
-    if git status | grep -q "modified:" ; then
-        VER="${VER}M"
-    fi
     if [ $LAVF == 1 ] ; then
         VER="${VER} [${BUILD_ARCH} ${BIT_DEPTH}bit-depth]"
     else
diff --git a/x264.c b/x264.c
index 1e49788..ad74ee5 100644
--- a/x264.c
+++ b/x264.c
@@ -30,8 +30,6 @@
 
 #ifdef _WIN32
 /* The following two defines must be located before the inclusion of any system header files. */
-#define WINVER       0x0500
-#define _WIN32_WINNT 0x0500
 #include <windows.h>
 #include <io.h>       /* _setmode() */
 #include <fcntl.h>    /* _O_BINARY */
