--- libtool-2.4.3/build-aux/ltmain.in.old	2014-10-28 15:15:09.143647100 +0900
+++ libtool-2.4.3/build-aux/ltmain.in	2014-10-28 15:16:02.113373400 +0900
@@ -6583,7 +6583,9 @@
 		  eval libdir=`$SED -n -e 's/^libdir=\(.*\)$/\1/p' $deplib`
 		  test -z "$libdir" && \
 		    func_fatal_error "'$deplib' is not a valid libtool archive"
-		  test "$absdir" != "$libdir" && \
+		  abs_inode=`ls -i "$deplib" | awk '{print $1}'`
+		    lib_inode=`ls -i "$libdir/$(basename $deplib)" | awk '{print $1}'`
+		  test "$abs_inode" != "$lib_inode" && \
 		    func_warning "'$deplib' seems to be moved"
 
 		  path=-L$absdir
