--- libtool-2.4.3/configure.old	2014-10-28 15:19:11.546024500 +0900
+++ libtool-2.4.3/configure	2014-10-28 15:19:39.817114500 +0900
@@ -29521,7 +29521,7 @@
 fi
 ac_cs_awk_cr=`$AWK 'BEGIN { print "a\rb" }' </dev/null 2>/dev/null`
 if test "$ac_cs_awk_cr" = "a${ac_cr}b"; then
-  ac_cs_awk_cr='\\r'
+  ac_cs_awk_cr='\r'
 else
   ac_cs_awk_cr=$ac_cr
 fi
--- libtool-2.4.3/libltdl/configure.old	2014-10-28 15:19:26.645441900 +0900
+++ libtool-2.4.3/libltdl/configure	2014-10-28 15:19:39.835116800 +0900
@@ -13971,7 +13971,7 @@
 fi
 ac_cs_awk_cr=`$AWK 'BEGIN { print "a\rb" }' </dev/null 2>/dev/null`
 if test "$ac_cs_awk_cr" = "a${ac_cr}b"; then
-  ac_cs_awk_cr='\\r'
+  ac_cs_awk_cr='\r'
 else
   ac_cs_awk_cr=$ac_cr
 fi
