From ef4b7cd33f1c6a7143070b1317e776d50b8fa996 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Fri, 29 May 2015 21:13:30 +0900
Subject: [PATCH 3/3] Use typedef instead of #define.

---
 zconf.h.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/zconf.h.in b/zconf.h.in
index e53866f..ca9c92c 100644
--- a/zconf.h.in
+++ b/zconf.h.in
@@ -92,8 +92,8 @@
 #endif
 
 /* Fallback for something that includes us. */
-#define Byte unsigned char
-#define Bytef unsigned char
+typedef unsigned char Byte;
+typedef Byte Bytef;
 
 typedef unsigned int   uInt;  /* 16 bits or more */
 typedef unsigned long  uLong; /* 32 bits or more */
-- 
2.4.1

