From 528e026eb329d894475b302206916eee7ef0ea9f Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sat, 2 May 2015 19:51:24 +0900
Subject: [PATCH 5/9] Add old typedef.

---
 zconf.h    | 16 ++++++++++++++++
 zconf.h.in | 16 ++++++++++++++++
 2 files changed, 32 insertions(+)

diff --git a/zconf.h b/zconf.h
index 6b05ec4..7e019c4 100644
--- a/zconf.h
+++ b/zconf.h
@@ -71,4 +71,20 @@
 
 typedef uint32_t z_crc_t;
 
+typedef unsigned char  Byte;
+typedef unsigned int   uInt;
+typedef unsigned long  uLong;
+typedef Byte           Bytef;
+typedef char           charf;
+typedef int            intf;
+typedef uInt           uIntf;
+typedef uLong          uLongf;
+typedef const void    *voidpc;
+typedef void          *voidpf;
+typedef void          *voidp;
+
+#ifndef Z_NULL
+#  define Z_NULL NULL
+#endif
+
 #endif /* ZCONF_H */
diff --git a/zconf.h.in b/zconf.h.in
index 6b05ec4..7e019c4 100644
--- a/zconf.h.in
+++ b/zconf.h.in
@@ -71,4 +71,20 @@
 
 typedef uint32_t z_crc_t;
 
+typedef unsigned char  Byte;
+typedef unsigned int   uInt;
+typedef unsigned long  uLong;
+typedef Byte           Bytef;
+typedef char           charf;
+typedef int            intf;
+typedef uInt           uIntf;
+typedef uLong          uLongf;
+typedef const void    *voidpc;
+typedef void          *voidpf;
+typedef void          *voidp;
+
+#ifndef Z_NULL
+#  define Z_NULL NULL
+#endif
+
 #endif /* ZCONF_H */
-- 
2.4.0

