From 6f4f669da5984646bad290d552effbc2fa3d8f53 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sat, 9 May 2015 01:54:45 +0900
Subject: [PATCH 8/8] Makefile.in: Add install-shared.

---
 Makefile.in | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/Makefile.in b/Makefile.in
index 67ae20a..3157414 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -194,6 +194,33 @@ example64$(EXE): example64.o $(STATICLIB)
 minigzip64$(EXE): minigzip64.o $(STATICLIB)
 	$(CC) $(CFLAGS) -o $@ minigzip64.o $(TEST_LDFLAGS)
 
+install-shared: $(SHAREDTARGET)
+	-@if [ ! -d $(DESTDIR)$(exec_prefix)  ]; then mkdir -p $(DESTDIR)$(exec_prefix); fi
+	-@if [ ! -d $(DESTDIR)$(libdir)       ]; then mkdir -p $(DESTDIR)$(libdir); fi
+	-@if [ ! -d $(DESTDIR)$(sharedlibdir) ]; then mkdir -p $(DESTDIR)$(sharedlibdir); fi
+	-@if [ ! -d $(DESTDIR)$(pkgconfigdir) ]; then mkdir -p $(DESTDIR)$(pkgconfigdir); fi
+ifneq ($(SHAREDTARGET),)
+ifneq ($(IMPORTLIB),)
+	cp $(IMPORTLIB) $(DESTDIR)$(sharedlibdir)
+	echo "cp $(IMPORTLIB) $(DESTDIR)$(sharedlibdir)"
+	chmod 644 $(DESTDIR)$(sharedlibdir)/$(IMPORTLIB)
+	-@if [ ! -d $(DESTDIR)$(bindir) ]; then mkdir -p $(DESTDIR)$(bindir); fi
+	cp $(SHAREDTARGET) $(DESTDIR)$(bindir)
+	echo "cp $(SHAREDTARGET) $(DESTDIR)$(bindir)"
+	chmod 755 $(DESTDIR)$(bindir)/$(SHAREDTARGET)
+	echo "chmod 755 $(DESTDIR)$(bindir)/$(SHAREDTARGET)"
+else
+	cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir)
+	echo "cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir)"
+	chmod 644 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET)
+	echo "chmod 644 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET)"
+	rm -f $(DESTDIR)$(sharedlibdir)/$(SHAREDLIB) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIBM)
+	ln -s $(SHAREDLIBV) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIB)
+	ln -s $(SHAREDLIBV) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIBM)
+	($(LDCONFIG) || true)  >/dev/null 2>&1
+endif
+endif
+
 install-libs: $(LIBS)
 	-@if [ ! -d $(DESTDIR)$(exec_prefix)  ]; then mkdir -p $(DESTDIR)$(exec_prefix); fi
 	-@if [ ! -d $(DESTDIR)$(libdir)       ]; then mkdir -p $(DESTDIR)$(libdir); fi
-- 
2.4.0

