From 2cb9839c572f47c48fe934ea4f9274a81717f797 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Tue, 5 May 2015 16:37:01 +0900
Subject: [PATCH 9/9] Makefile.in: Add install-shared.

---
 Makefile.in | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/Makefile.in b/Makefile.in
index 3e75774..c9f6fe3 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -227,6 +227,43 @@ install-libs: $(LIBS)
 # The ranlib in install is needed on NeXTSTEP which checks file times
 # ldconfig is for Linux
 
+install-shared: $(SHAREDTARGET)
+	-@if [ ! -d $(DESTDIR)$(exec_prefix)  ]; then mkdir -p $(DESTDIR)$(exec_prefix); fi
+	-@if [ ! -d $(DESTDIR)$(libdir)       ]; then mkdir -p $(DESTDIR)$(libdir); fi
+	-@if [ ! -d $(DESTDIR)$(sharedlibdir) ]; then mkdir -p $(DESTDIR)$(sharedlibdir); fi
+	-@if [ ! -d $(DESTDIR)$(pkgconfigdir) ]; then mkdir -p $(DESTDIR)$(pkgconfigdir); fi
+	-@if test -n "$(IMPORTLIB)" ; then \
+		cp $(IMPORTLIB) $(DESTDIR)$(libdir); \
+		echo "cp $(IMPORTLIB) $(DESTDIR)$(libdir)"; \
+		chmod 644 $(DESTDIR)$(libdir)/$(IMPORTLIB); \
+		if test -n "$(SHAREDTARGET)" ; then \
+			if test -f "$(SHAREDTARGET)" ; then \
+				cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir); \
+				echo "cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir)"; \
+				chmod 755 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET); \
+				echo "chmod 755 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET)"; \
+			fi; \
+		fi; \
+		if test -n "$(SHAREDLIBV)"; then \
+			if test -f "$(SHAREDLIBV)"; then \
+				rm -f $(DESTDIR)$(sharedlibdir)/$(SHAREDLIB) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIBM); \
+				ln -s $(SHAREDLIBV) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIB); \
+				ln -s $(SHAREDLIBV) $(DESTDIR)$(sharedlibdir)/$(SHAREDLIBM); \
+				($(LDCONFIG) || true)  >/dev/null 2>&1; \
+			fi; \
+		fi; \
+	else \
+		cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir); \
+		echo "cp $(SHAREDTARGET) $(DESTDIR)$(sharedlibdir)"; \
+		chmod 755 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET); \
+		echo "chmod 755 $(DESTDIR)$(sharedlibdir)/$(SHAREDTARGET)"; \
+	fi
+	cp zlib.pc $(DESTDIR)$(pkgconfigdir)
+	chmod 644 $(DESTDIR)$(pkgconfigdir)/zlib.pc
+	-@if [ ! -d $(DESTDIR)$(includedir)   ]; then mkdir -p $(DESTDIR)$(includedir); fi
+	cp $(SRCDIR)/zlib.h zconf.h $(DESTDIR)$(includedir)
+	chmod 644 $(DESTDIR)$(includedir)/zlib.h $(DESTDIR)$(includedir)/zconf.h
+
 install: install-libs
 	-@if [ ! -d $(DESTDIR)$(includedir)   ]; then mkdir -p $(DESTDIR)$(includedir); fi
 	cp $(SRCDIR)/zlib.h zconf.h $(DESTDIR)$(includedir)
-- 
2.4.0

