From 7a1bd1d46bad9a94a3cb381604a3557a9a6c9e3d Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Fri, 15 Apr 2011 18:09:48 +0200
Subject: [PATCH 02/60] HACK: Disable tb_unreliable, preserve original fps
 information.

---
 libavformat/utils.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/libavformat/utils.c b/libavformat/utils.c
index 75dba9b..1b0dcd5 100644
--- a/libavformat/utils.c
+++ b/libavformat/utils.c
@@ -2945,14 +2945,14 @@ static int get_std_framerate(int i)
 static int tb_unreliable(AVCodecContext *c)
 {
     if (c->time_base.den >= 101LL * c->time_base.num ||
-        c->time_base.den <    5LL * c->time_base.num ||
-        // c->codec_tag == AV_RL32("DIVX") ||
-        // c->codec_tag == AV_RL32("XVID") ||
+        c->time_base.den <    5LL * c->time_base.num/* ||
+        c->codec_tag == AV_RL32("DIVX") ||
+        c->codec_tag == AV_RL32("XVID") ||
         c->codec_tag == AV_RL32("mp4v") ||
         c->codec_id == AV_CODEC_ID_MPEG2VIDEO ||
         c->codec_id == AV_CODEC_ID_GIF ||
         c->codec_id == AV_CODEC_ID_HEVC ||
-        c->codec_id == AV_CODEC_ID_H264)
+        c->codec_id == AV_CODEC_ID_H264*/)
         return 1;
     return 0;
 }
-- 
2.5.2

