From c4ddf5feb4dafca27f2e84f705db8c1f7eaa96af Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Tue, 6 Aug 2013 20:24:31 +0200
Subject: [PATCH 20/65] riff: don't try to parse too small WAVEFORMAT headers

---
 libavformat/riffdec.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/libavformat/riffdec.c b/libavformat/riffdec.c
index f44df1e..43a8913 100644
--- a/libavformat/riffdec.c
+++ b/libavformat/riffdec.c
@@ -85,8 +85,12 @@ int ff_get_wav_header(AVIOContext *pb, AVCodecContext *codec, int size, int big_
 {
     int id;
 
-    if (size < 14)
+    /* avoid reading headers which are way too small */
+    if (size < 14) {
         avpriv_request_sample(codec, "wav header size < 14");
+        avio_skip(pb, size);
+        return 0;
+    }
 
     codec->codec_type  = AVMEDIA_TYPE_AUDIO;
     if (!big_endian) {
-- 
2.3.5

