From ef862eb3c2a29c16313fd27560b877e66ee360ce Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Wed, 23 Oct 2013 09:18:16 +0200
Subject: [PATCH 24/62] riff: add HEVC tags

---
 libavformat/riff.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libavformat/riff.c b/libavformat/riff.c
index f297dd4..694021b 100644
--- a/libavformat/riff.c
+++ b/libavformat/riff.c
@@ -377,6 +377,9 @@ const AVCodecTag ff_codec_bmp_tags[] = {
     { AV_CODEC_ID_SCREENPRESSO, MKTAG('S', 'P', 'V', '1') },
     { AV_CODEC_ID_RSCC,         MKTAG('R', 'S', 'C', 'C') },
     { AV_CODEC_ID_RSCC,         MKTAG('I', 'S', 'C', 'C') },
+    { AV_CODEC_ID_HEVC,         MKTAG('H', 'E', 'V', 'C') },
+    { AV_CODEC_ID_HEVC,         MKTAG('H', 'V', 'C', '1') },
+    { AV_CODEC_ID_HEVC,         MKTAG('H', 'M', '1', '0') },
     { AV_CODEC_ID_NONE,         0 }
 };
 
-- 
2.6.3

