From eca6ea90c08b71585256713d7b21ea659810f66b Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Wed, 23 Oct 2013 09:53:39 +0200
Subject: [PATCH 25/62] HACK: hevc: ignore invalid extradata

---
 libavcodec/hevc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libavcodec/hevc.c b/libavcodec/hevc.c
index ece36f8..5cf38d9 100644
--- a/libavcodec/hevc.c
+++ b/libavcodec/hevc.c
@@ -3263,8 +3263,8 @@ static av_cold int hevc_decode_init(AVCodecContext *avctx)
     if (avctx->extradata_size > 0 && avctx->extradata) {
         ret = hevc_decode_extradata(s);
         if (ret < 0) {
-            hevc_decode_free(avctx);
-            return ret;
+            s->is_nalff = 0;
+            av_log(avctx, AV_LOG_ERROR, "Invalid extradata ignored\n");
         }
     }
 
-- 
2.6.3

