From 85665cd7968a2b7498d1b3fda1f6bf06def692f5 Mon Sep 17 00:00:00 2001
From: Carl Eugen Hoyos <cehoyos@ag.or.at>
Date: Tue, 1 Apr 2014 16:00:23 +0200
Subject: [PATCH 33/62] riff: add ProRes FourCCs

---
 libavformat/riff.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/libavformat/riff.c b/libavformat/riff.c
index 8fa22e9..ef2ca89 100644
--- a/libavformat/riff.c
+++ b/libavformat/riff.c
@@ -363,6 +363,11 @@ const AVCodecTag ff_codec_bmp_tags[] = {
     { AV_CODEC_ID_FIC,          MKTAG('F', 'I', 'C', 'V') },
     { AV_CODEC_ID_PRORES,       MKTAG('A', 'P', 'C', 'N') },
     { AV_CODEC_ID_PRORES,       MKTAG('A', 'P', 'C', 'H') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'h') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'n') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 's') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'o') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', '4', 'h') },
     { AV_CODEC_ID_QTRLE,        MKTAG('r', 'l', 'e', ' ') },
     { AV_CODEC_ID_HEVC,         MKTAG('H', 'E', 'V', 'C') },
     { AV_CODEC_ID_HEVC,         MKTAG('H', 'V', 'C', '1') },
-- 
2.3.0

