From ba4ccc2f13a6e17641275a816954209f288677e4 Mon Sep 17 00:00:00 2001
From: Carl Eugen Hoyos <cehoyos@ag.or.at>
Date: Tue, 1 Apr 2014 16:00:23 +0200
Subject: [PATCH 33/65] riff: add ProRes FourCCs

---
 libavformat/riff.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/libavformat/riff.c b/libavformat/riff.c
index fbd2cb0..fea9452 100644
--- a/libavformat/riff.c
+++ b/libavformat/riff.c
@@ -369,6 +369,11 @@ const AVCodecTag ff_codec_bmp_tags[] = {
     { AV_CODEC_ID_HEVC,         MKTAG('H', 'E', 'V', 'C') },
     { AV_CODEC_ID_HEVC,         MKTAG('H', 'V', 'C', '1') },
     { AV_CODEC_ID_HEVC,         MKTAG('H', 'M', '1', '0') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'h') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'n') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 's') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', 'c', 'o') },
+    { AV_CODEC_ID_PRORES,       MKTAG('a', 'p', '4', 'h') },
     { AV_CODEC_ID_NONE,         0 }
 };
 
-- 
2.3.5

