From 039e27b774a420581b0fefca7bff29b6a6d3e7b2 Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Thu, 25 Sep 2014 20:02:47 +0200
Subject: [PATCH 37/66] hevc: scan for sps/pps/vps more aggressively

---
 libavcodec/hevc_parser.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/libavcodec/hevc_parser.c b/libavcodec/hevc_parser.c
index 3c7c3c3..d38376a 100644
--- a/libavcodec/hevc_parser.c
+++ b/libavcodec/hevc_parser.c
@@ -299,8 +299,6 @@ static int hevc_split(AVCodecContext *avctx, const uint8_t *buf, int buf_size)
             has_ps = 1;
         else if (has_ps)
             return ptr - 4 - buf;
-        else // no parameter set at the beginning of the stream
-            return 0;
     }
     return 0;
 }
-- 
2.4.3

