From 925ddd69b08f00931de976da143fac54c3e020fc Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Tue, 2 Dec 2014 14:44:36 +0100
Subject: [PATCH 40/65] mpeg12dec: properly handle a codec_id change

---
 libavcodec/mpeg12dec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libavcodec/mpeg12dec.c b/libavcodec/mpeg12dec.c
index a5622bf..89aa028 100644
--- a/libavcodec/mpeg12dec.c
+++ b/libavcodec/mpeg12dec.c
@@ -61,6 +61,7 @@ typedef struct Mpeg1Context {
     int slice_count;
     AVRational save_aspect;
     int save_width, save_height, save_progressive_seq;
+    enum AVCodecID save_codec_id;
     AVRational frame_rate_ext;  /* MPEG-2 specific framerate modificator */
     int sync;                   /* Did we reach a sync point like a GOP/SEQ/KEYFrame? */
     int tmpgexs;
@@ -1324,6 +1325,7 @@ static int mpeg_decode_postinit(AVCodecContext *avctx)
         s1->save_height          != s->height               ||
         av_cmp_q(s1->save_aspect, s->avctx->sample_aspect_ratio) ||
         (s1->save_progressive_seq != s->progressive_sequence && FFALIGN(s->height, 16) != FFALIGN(s->height, 32)) ||
+        s1->save_codec_id        != s->codec_id             ||
         0) {
         if (s1->mpeg_enc_ctx_allocated) {
             ParseContext pc = s->parse_context;
@@ -1347,6 +1349,7 @@ static int mpeg_decode_postinit(AVCodecContext *avctx)
         s1->save_width           = s->width;
         s1->save_height          = s->height;
         s1->save_progressive_seq = s->progressive_sequence;
+        s1->save_codec_id        = s->codec_id;
 
         /* low_delay may be forced, in this case we will have B-frames
          * that behave like P-frames. */
@@ -2240,6 +2243,7 @@ static int vcr2_init_sequence(AVCodecContext *avctx)
     s1->save_width           = s->width;
     s1->save_height          = s->height;
     s1->save_progressive_seq = s->progressive_sequence;
+    s1->save_codec_id        = s->codec_id;
     return 0;
 }
 
-- 
2.3.5

