From 1250efbf4a8cf10fd5166d1f06475d001bffdc35 Mon Sep 17 00:00:00 2001
From: Hendrik Leppkes <h.leppkes@gmail.com>
Date: Wed, 18 Mar 2015 11:11:10 +0100
Subject: [PATCH 40/63] dca_parser: don't overwrite the sample rate, it may not
 be correct

---
 libavcodec/dca_parser.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/libavcodec/dca_parser.c b/libavcodec/dca_parser.c
index c95fedb..7d7868f 100644
--- a/libavcodec/dca_parser.c
+++ b/libavcodec/dca_parser.c
@@ -193,7 +193,6 @@ static int dca_parse(AVCodecParserContext *s, AVCodecContext *avctx,
     /* read the duration and sample rate from the frame header */
     if (!dca_parse_params(buf, buf_size, &duration, &sample_rate, &pc1->framesize)) {
         s->duration        = duration;
-        avctx->sample_rate = sample_rate;
     } else
         s->duration = 0;
 
-- 
2.5.0

