From a42108985823880c57ccd481b19c8e27ae524b2b Mon Sep 17 00:00:00 2001
From: Tsukasa OMOTO <henry0312@gmail.com>
Date: Fri, 4 May 2012 02:32:46 +0900
Subject: [PATCH 45/66] isom: Support demuxing more DTS audio in ISOBMFF.

---
 libavformat/isom.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/libavformat/isom.c b/libavformat/isom.c
index c99f2dc..90f9b18 100644
--- a/libavformat/isom.c
+++ b/libavformat/isom.c
@@ -270,6 +270,7 @@ const AVCodecTag ff_codec_movaudio_tags[] = {
     { AV_CODEC_ID_AMR_NB,          MKTAG('s', 'a', 'm', 'r') }, /* AMR-NB 3gp */
     { AV_CODEC_ID_AMR_WB,          MKTAG('s', 'a', 'w', 'b') }, /* AMR-WB 3gp */
     { AV_CODEC_ID_DTS,             MKTAG('d', 't', 's', 'c') }, /* DTS formats prior to DTS-HD */
+    { AV_CODEC_ID_DTS,             MKTAG('d', 't', 's', 'e') }, /* DTS Low Bit Rate (LBR) formats */
     { AV_CODEC_ID_DTS,             MKTAG('d', 't', 's', 'h') }, /* DTS-HD audio formats */
     { AV_CODEC_ID_DTS,             MKTAG('d', 't', 's', 'l') }, /* DTS-HD Lossless formats */
     { AV_CODEC_ID_DTS,             MKTAG('D', 'T', 'S', ' ') }, /* non-standard */
-- 
2.4.1

