From 48ab408d75bebfda3dfb8a5faa304f39b69cf2fc Mon Sep 17 00:00:00 2001
From: Yusuke Nakamura <muken.the.vfrmaniac@gmail.com>
Date: Sat, 8 Nov 2014 18:30:19 +0900
Subject: [PATCH 46/62] alsdec: Fix avoiding reading over the end of packet
 data.

---
 libavcodec/alsdec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libavcodec/alsdec.c b/libavcodec/alsdec.c
index ebd364e..78e63d1 100644
--- a/libavcodec/alsdec.c
+++ b/libavcodec/alsdec.c
@@ -1515,6 +1515,10 @@ static int decode_frame(AVCodecContext *avctx, void *data, int *got_frame_ptr,
     int invalid_frame, ret;
     unsigned int c, sample, ra_frame, bytes_read, shift;
 
+    /* After seek to the non-first frame, we can't get the last frame length correctly.
+     * This will avoid reading over the end of packet data. */
+    memset(avpkt->data + avpkt->size, 0, FF_INPUT_BUFFER_PADDING_SIZE);
+
     if ((ret = init_get_bits8(&ctx->gb, buffer, buffer_size)) < 0)
         return ret;
 
-- 
2.6.3

