From 2aa85b063d147ebbfb8b08b2773981d6da29c7f3 Mon Sep 17 00:00:00 2001
From: Yusuke Nakamura <muken.the.vfrmaniac@gmail.com>
Date: Sat, 8 Nov 2014 18:30:19 +0900
Subject: [PATCH 50/66] alsdec: Fix avoiding reading over the end of packet
 data.

---
 libavcodec/alsdec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libavcodec/alsdec.c b/libavcodec/alsdec.c
index 75be55c..14fc13a 100644
--- a/libavcodec/alsdec.c
+++ b/libavcodec/alsdec.c
@@ -1510,6 +1510,10 @@ static int decode_frame(AVCodecContext *avctx, void *data, int *got_frame_ptr,
     int invalid_frame, ret;
     unsigned int c, sample, ra_frame, bytes_read, shift;
 
+    /* After seek to the non-first frame, we can't get the last frame length correctly.
+     * This will avoid reading over the end of packet data. */
+    memset(avpkt->data + avpkt->size, 0, FF_INPUT_BUFFER_PADDING_SIZE);
+
     if ((ret = init_get_bits8(&ctx->gb, buffer, buffer_size)) < 0)
         return ret;
 
-- 
2.4.1

