From b9a86f6b65a91a05c2cb49fa46549887a847e17e Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 15 Sep 2013 10:35:05 +0900
Subject: [PATCH 51/63] configure: Change mingw32* into mingw*.

---
 configure | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/configure b/configure
index 1f9a1b7..82d4f52 100755
--- a/configure
+++ b/configure
@@ -3290,7 +3290,7 @@ fi
 
 exesuf() {
     case $1 in
-        mingw32*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
+        mingw*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
     esac
 }
 
@@ -4335,11 +4335,11 @@ case $target_os in
         enabled_any pic shared x86_64 ||
             { check_cflags -mdynamic-no-pic && add_asflags -mdynamic-no-pic; }
         ;;
-    mingw32*)
+    mingw*)
         if test $target_os = "mingw32ce"; then
             disable network
         else
-            target_os=mingw32
+            target_os=mingw
         fi
         decklink_outdev_extralibs="$decklink_outdev_extralibs -lole32 -loleaut32"
         decklink_indev_extralibs="$decklink_indev_extralibs -lole32 -loleaut32"
-- 
2.5.0

