From a0ff0db750c0322eacb366ca7bea297ce750f3c4 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sun, 15 Sep 2013 10:35:05 +0900
Subject: [PATCH 53/65] configure: Change mingw32* into mingw*.

---
 configure | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/configure b/configure
index 8f8aeef..a5c7beb 100755
--- a/configure
+++ b/configure
@@ -3152,7 +3152,7 @@ fi
 
 exesuf() {
     case $1 in
-        mingw32*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
+        mingw*|win32|win64|cygwin*|*-dos|freedos|opendos|os/2*|symbian) echo .exe ;;
     esac
 }
 
@@ -4163,11 +4163,11 @@ case $target_os in
         enabled_any pic shared x86_64 ||
             { check_cflags -mdynamic-no-pic && add_asflags -mdynamic-no-pic; }
         ;;
-    mingw32*)
+    mingw*)
         if test $target_os = "mingw32ce"; then
             disable network
         else
-            target_os=mingw32
+            target_os=mingw
         fi
         LIBTARGET=i386
         if enabled x86_64; then
-- 
2.3.5

