From 399811e1552a4667a1222d0596775c983697f187 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sat, 10 Jan 2015 18:02:19 +0900
Subject: [PATCH 56/60] libutvideodec: Fix bits_per_raw_sample.

---
 libavcodec/libutvideodec.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/libavcodec/libutvideodec.cpp b/libavcodec/libutvideodec.cpp
index 47261a6..d91201a 100644
--- a/libavcodec/libutvideodec.cpp
+++ b/libavcodec/libutvideodec.cpp
@@ -108,8 +108,14 @@ static av_cold int utvideo_decode_init(AVCodecContext *avctx)
     /* Allocate the output frame */
     avctx->coded_frame = av_frame_alloc();
 
-    /* Ut Video only supports 8-bit */
+#ifdef UTVF_UQY2
+    if (format == UTVF_v210)
+        avctx->bits_per_raw_sample = 10;
+    else
+        avctx->bits_per_raw_sample = 8;
+#else
     avctx->bits_per_raw_sample = 8;
+#endif
 
     /* Is it interlaced? */
     avctx->coded_frame->interlaced_frame = info.flags & 0x800 ? 1 : 0;
-- 
2.5.2

