From a9981a0ea78d743e6671c03d4864844e0d9636c1 Mon Sep 17 00:00:00 2001
From: Yuta NAKAI <nak5124@live.jp>
Date: Sat, 30 May 2015 07:04:05 +0900
Subject: [PATCH 60/62] ffmpeg: Don't show banner by default.

---
 cmdutils.c             | 6 +++---
 cmdutils_common_opts.h | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/cmdutils.c b/cmdutils.c
index 41daa95..dc7a614 100644
--- a/cmdutils.c
+++ b/cmdutils.c
@@ -69,7 +69,7 @@ AVDictionary *format_opts, *codec_opts, *resample_opts;
 
 static FILE *report_file;
 static int report_file_level = AV_LOG_DEBUG;
-int hide_banner = 0;
+int hide_banner = 1;
 
 void init_opts(void)
 {
@@ -508,9 +508,9 @@ void parse_loglevel(int argc, char **argv, const OptionDef *options)
             fflush(report_file);
         }
     }
-    idx = locate_option(argc, argv, options, "hide_banner");
+    idx = locate_option(argc, argv, options, "show_banner");
     if (idx)
-        hide_banner = 1;
+        hide_banner = 0;
 }
 
 static const AVOption *opt_find(void *obj, const char *name, const char *unit,
diff --git a/cmdutils_common_opts.h b/cmdutils_common_opts.h
index 758dac1..2a10cbb 100644
--- a/cmdutils_common_opts.h
+++ b/cmdutils_common_opts.h
@@ -22,7 +22,7 @@
     { "report"     , 0,        {(void*)opt_report}, "generate a report" },
     { "max_alloc"  , HAS_ARG,  {.func_arg = opt_max_alloc},     "set maximum size of a single allocated block", "bytes" },
     { "cpuflags"   , HAS_ARG | OPT_EXPERT, { .func_arg = opt_cpuflags }, "force specific cpu flags", "flags" },
-    { "hide_banner", OPT_BOOL | OPT_EXPERT, {&hide_banner},     "do not show program banner", "hide_banner" },
+    { "show_banner", OPT_BOOL | OPT_EXPERT, {&hide_banner},     "show program banner", "show_banner" },
 #if CONFIG_OPENCL
     { "opencl_bench", OPT_EXIT, {.func_arg = opt_opencl_bench}, "run benchmark on all OpenCL devices and show results" },
     { "opencl_options", HAS_ARG, {.func_arg = opt_opencl},      "set OpenCL environment options" },
-- 
2.6.3

