diff --git a/libavcodec/libutvideoenc.cpp b/libavcodec/libutvideoenc.cpp
index 3b88fff..6d541bb 100644
--- a/libavcodec/libutvideoenc.cpp
+++ b/libavcodec/libutvideoenc.cpp
@@ -120,9 +120,14 @@ static av_cold int utvideo_encode_init(AVCodecContext *avctx)
 
     /* Get extradata from encoder */
     avctx->extradata_size = utv->codec->EncodeGetExtraDataSize();
+#if UTVIDEO_VERSION_MAJOR > 16 || (UTVIDEO_VERSION_MAJOR == 15 && UTVIDEO_VERSION_MINOR >= 2)
+    utv->codec->EncodeGetExtraData(info, avctx->extradata_size, in_format,
+                                   avctx->width, avctx->height);
+#else
     utv->codec->EncodeGetExtraData(info, avctx->extradata_size, in_format,
                                    avctx->width, avctx->height,
                                    CBGROSSWIDTH_WINDOWS);
+#endif
     avctx->extradata = (uint8_t *)info;
 
     /* Set flags */
diff --git a/libavcodec/libutvideodec.cpp b/libavcodec/libutvideodec.cpp
index e4b87a8..ac93076 100644
--- a/libavcodec/libutvideodec.cpp
+++ b/libavcodec/libutvideodec.cpp
@@ -150,7 +156,7 @@ static int utvideo_decode_frame(AVCodecContext *avctx, void *data,
     pic->key_frame = 1;
 
     /* Decode the frame */
-    utv->codec->DecodeFrame(utv->buffer, avpkt->data, true);
+    utv->codec->DecodeFrame(utv->buffer, avpkt->data);
 
     /* Set the output data depending on the colorspace */
     switch (avctx->pix_fmt) {
